Proposal: Variadics (#2240)
carbon-language/carbon-lang: proposals/p2240.md

[performance-critical software](/docs/project/goals.md#performance-critical-software),
because variadic APIs can be more efficient than their non-variadic
counterparts. For example, `StrCat` is fundamentally more efficient than
something like a chain of `operator+` calls on `std::string`, because it does
not need to materialize a series of partial results, and it can pre-allocate a
buffer large enough for the final result.

This proposal deviates somewhat from the principle that
[all APIs are library APIs](/docs/project/principles/library_apis_only.md)
because it treats tuples and packs as effectively built-in. As the document for
this principle notes, tuples may be an exception to the principle because it is
very hard to see how we could define them as library types without circularity.
Packs have much the same problem, since a type representing a pack would itself
have to be a variadic type. The need for both types to deeply support symbolic
computation adds some major challenges here as well. However, if tuples could
ever be library types, they would certainly need to be defined variadically, so
in that sense this proposal may actually support that principle.

> 

Reviewer: Issue: performance of generated code.

The more I've thought about this, the more I think we should take this consideration off the table unless we have some evidence that one approach would be more efficient. My suspicion is that we can make a ABI/calling convention that is efficient for all of the different alternatives:
* If the arguments do not have the same type, monomorphize.
* Otherwise pass all the arguments in order contiguously on the stack along with the total number. This should allow a single implementation to work with a variety of arities, and doesn't care how you have written the parameter signature.

Author: Is this comment still applicable or can it be resolved?

> ## Alternatives considered

Reviewer: Issue: restrictive matching of symbolic packs to parameter lists.

The more I think about this option, the more I dislike it:
* It creates an unnecessary difference between the template and symbolic cases at the call boundary, opposite the direction of #2153 .
* It is an additional rule to explain to developers, and the rule needs to be pretty subtle in order to allow things like forwarding and destructuring.
* I'm pretty worried that a consequence of this subtle rule is that we are going to reject things that developers will expect to work and it will be really hard to give an error explaining what they did wrong.
* It means there is additional state associated with a function declaration around how the arguments are declared that needs to be modeled somehow. I'm still working out how much of an issue this is, but for example the things I'm thinking about are how this interacts with #2875 and functions actually being implementations of the `Call` interface for a type.

Furthermore, I feel like the alternative is pretty good: there are cases that we have to reject, but in those cases there is a pretty clear error we can give ("no single return type could be deduced", or whatever).

> 

Reviewer: Issue: `each(>=1)` syntax

This issue I'm more neutral on, but I think it is extra complexity we don't need if we don't have restrictive matching. It means that there is additional syntax to learn, more ways to write function signatures, and those ways are subtly different in ways that are hard to explain. It also has concerns around which `each` needs to be marked with `(>=1)` and extra round trips if the user gets it wrong.

Reviewer: Perhaps include this rationale in the "No parameter merging" alternative?

Author: > It means that there is additional syntax to learn, more ways to write function signatures,

OK, I've added those points.

> and those ways are subtly different in ways that are hard to explain. 

I don't agree with this one:

```
fn F[A:! I, ... each B:! I](a: A, ... each b: each B);
fn G[A:! I, ... each B:! I](... each b: each B, a: A);
fn H[A:! I, ... each B:! I](... each b: each B, a: A) -> A;
```

Under the primary proposal, `F` and `G` are equivalent (in terms of the requirements they impose on the argument list), but `G` and `H` are different. Under this alternative, `F` and `G` are different, but `G` and `H` are equivalent. I don't know how to convincingly argue that the latter state of affairs is any more subtle or hard to explain than the former. If anything, it seems easier to explain, because there's a direct (but perhaps unexpected) connection between the signature syntax and the callsite requirements.

> It also has concerns around which `each` needs to be marked with `(>=1)` and extra round trips if the user gets it wrong.

This seems like a corollary of the fact that `(>=1)` conceptually belongs on `...` rather than `each`, which is already discussed in the rationale.

(with a disjoint syntax).
[SE-0404: Pack Iteration](https://github.com/simanerush/swift-evolution/blob/se-0404-pack-iteration/proposals/0404-pack-iteration.md),
which extends that to support iterating through a variadic parameter list, has
been positively received, but hasn't yet been approved.

There have been several attempts to add such a feature to Rust, but that work is
[currently inactive](https://github.com/rust-lang/rfcs/issues/376#issuecomment-830034029).

## Proposal

See `/docs/design/variadics.md` in this pull request.

### Examples

```carbon
// Computes the sum of its arguments, which are i64s
fn SumInts(... each param: i64) -> i64 {
  var sum: i64 = 0;
  ... sum += each param;
  return sum;
}
```

```carbon
// Concatenates its arguments, which are all convertible to String
fn StrCat[... each T:! ConvertibleToString](... each param: each T) -> String {
  var len: i64 = 0;
  ... len += each param.Length();
  var result: String = "";
  result.Reserve(len);
  ... result.Append(each param.ToString());
  return result;
}
```

```carbon
// Returns the minimum of its arguments, which must all have the same type T.

> fn Min[T:! Comparable & Value](... each(>=1) param: T) -> T {
>   let (var result: T, ... each next: T) = (... each param);

Reviewer: Should the `each(>=1)` syntax have been updated?
```suggestion
fn Min[T:! Comparable & Value](var result: T, ... each next: T) -> T {
```

  ... var each iter: auto = each vector.Begin();
  var result: Vector((... each ElementType));
  while (...and each iter != each vector.End()) {
    result.push_back((... each iter));
    ... each iter++;
  }
  return result;
}
```

```carbon
// Toy example of mixing variadic and non-variadic parameters.
// Takes an i64, any number of f64s, and then another i64.
fn MiddleVariadic(first: i64, ... each middle: f64, last: i64);
```

```carbon
// Toy example of using the result of variadic type deduction.
fn TupleConcat[... each T1: type, ... each T2: type](
    t1: (... each T1), t2: (... each T2)) -> (... each T1, ... each T2) {
  return (...expand t1, ...expand t2);
}
```

### Comparisons

The following table compares selected examples of Carbon variadics against
equivalent code written in C++20 (with and without the extensions discussed
[earlier](#background)) and Swift.

<table>
<tr>
<th>Carbon</th>
<th>C++20</th>
<th>C++20 with extensions</th>
<th>Swift with extensions</th>
</tr>

<tr>
<td>

```carbon
// Computes the sum of its arguments, which are i64s
fn SumInts(... each param: i64) -> i64 {
  var sum: i64 = 0;
  ... sum += each param;
  return sum;
}
```

</td>
<td>

> ```C++

Reviewer 2: I think we've been using `cpp` to match our file extension in most places?

````suggestion
```cpp
````

template <typename... Params>
    requires (std::convertible_to<Params, int64_t> && ...)
int64_t SumInts(const Params&... params) {
  return (static_cast<int64_t>(params) + ... + 0);
}
```

</td>
<td>

With P1219R2:

```C++
int64_t SumInts(int64... params) {
  return (static_cast<int64_t>(params) + ... + 0);
}
```

</td>
<td>

(No extensions)

```swift
func SumInts(_ params: Int64...) {
  var sum: Int64 = 0
  for param in params {
    sum += param
  }
  return sum
}
```

</td>
</tr>
<tr>
<td>

```carbon

> fn Min[T:! Comparable & Value](... each(>=1) param: T) -> T {

Reviewer: Here too.

  let (var result: T, ... each next: T) = (... each param);
  ... if (each next < result) {
    result = each next;
  }
  return result;
}
```

</td>
<td>

> ```c++

Reviewer 2: ````suggestion
```cpp
````

template <typename T>
    requires std::totally_ordered<T> && std::copyable<T>
T Min(const T& t) {
  return t;
}

template <typename T, typename... Params>
    requires std::totally_ordered<T> && std::copyable<T> &&
             (std::same_as<T, Params> && ...)
T Min(T first, Params... rest) {
  T min_rest = Min(rest...);
  if (min_rest < first) {
    return min_rest;
  } else {
    return first;
  }

> }

Reviewer 2: FWIW, I think it might be worth rewriting the C++ examples to use `if constexpr` to provide the base case without overloading as that seems a bit briefer and more representative, here and below. WDYT?

Here is what I think this one would look like:

```suggestion
template <typename T, typename... Params>
    requires std::totally_ordered<T> && std::copyable<T> &&
             (std::same_as<T, Params> && ...)
T Min(T first, Params... rest) {
  // Base case.
  if constexpr (sizeof...(rest) == 0) {
    return first;
  } else {
    T min_rest = Min(rest...);
    if (min_rest < first) {
      return min_rest;
    } else {
      return first;
    }
  }
}
```

Reviewer: I would put the `// Base case.` comment inside the first branch of the `if`.

I feel very mixed about this suggestion since the overloading approach was the way to do this in the day, but I guess C++ has moved and the example is documented as using C++20.

Author: As far as I can tell this technique only works if the minimum arity is at least 1, so that the signature can do the destructuring while still supporting all usages. That makes it a great fit for `Min`, but awkward to apply to `StrCat`. I've done it (see below), but I had to rely on the fact that there's a separate implementation function (which is basically a coincidence), and the result still doesn't seem like a net improvement to me.

Reviewer: I don't think this `if constexpr` trick is an improvement for `StrCat`, where it creates two edge cases instead of one. @zygoloid had an alternate implementation that uses fold expressions to avoid quadratic compile time and the need for a helper function, but I don't know if that serves the narrative purpose you are looking for. Is this supposed to be a common way of writing this function, a way that shows common pitfalls, or a way that is as similar as possible to the Carbon approach?

Author: I had been aiming for these examples to be representative of how functions like this would typically be written. I think that means we shouldn't use the comma-fold trick, and probably not `if constexpr` either, although that's more debatable. However, it sounds like maybe Chandler is asking for the examples to represent the ideal versions of these functions; in that case, it might make sense to use the comma-fold trick in `StrCat`. In either case, I agree with you that `StrCat` should not use `if constexpr`, but I'll wait for @chandlerc to clarify what he'd like to see.

Author: In an offline conversation, Chandler indicated he thinks the C++ examples should be as elegant and optimistic as we can make them, unless we're specifically focusing on and discussing the issue that typical C++ code is suboptimal in some way. We're not focusing on that kind of issue here, so I've rewritten this to the comma-fold version.

(Chandler, feel free to clarify/correct my paraphrase of what you said).

Reviewer 2: Looks good to me!

```

</td>
<td>

With P1219R2 and P1306R2

> ```c++

Reviewer 2: ````suggestion
```cpp
````

template <typename T>
    requires std::totally_ordered<T> && std::copyable<T>
T Min(const T& first, const T&... rest) {
  T result = first;
  template for (const T& t: rest) {
    if (t < result) {
      result = t;
    }
  }
  return result;
}
```

</td>
<td>

(No extensions)

```swift
func Min<T: Comparable>(_ first: T, _ rest: T...) -> T {
  var result: T = first;
  for t in rest {
    if (t < result) {
      result = t
    }
  }
  return result
}
```

</td>
</tr>
<tr>
<td>

```carbon
fn StrCat[... each T:! ConvertibleToString](... each param: each T) -> String {
  var len: i64 = 0;
  ... len += each param.Length();
  var result: String = "";
  result.Reserve(len);
  ... result.Append(each param.ToString());
  return result;
}
```

</td>
<td>

> ```c++

Reviewer 2: ````suggestion
```cpp
````

template <ConvertibleToString... Ts>
std::string StrCat(const Ts&... params) {
  std::string result;
  result.reserve((params.Length() + ... + 0));
  StrCatImpl(&result, params...);
  return result;
}

void StrCatImpl(std::string* out) { return; }

template <ConvertibleToString T, ConvertibleToString... Ts>
void StrCatImpl(std::string* out, const T& first, const Ts&... rest) {
  out->append(first.ToString());
  StrCatImpl(out, rest...);
}
```

</td>
<td>

With P1306R2

> ```c++

Reviewer 2: ````suggestion
```cpp
````

occur in ordinary binary syntax has few advantages and several drawbacks:

-   It might conflict with a future general fold facility.
-   It would invite users to try other operators, and would probably give less
    clear errors if they do.
-   It would substantially complicate parsing and the AST.
-   It would force users to make a meaningless choice between `x or ...` and
    `... or x`, and likewise for `and`.

See also the discussion [below](#fold-like-syntax) of using `...,` and `...;` in
place of the tuple and statement forms of `...`. This is inspired by fold
expressions, but distinct from them, because `,` and `;` are not truly binary
operators, and it's targeting a different problem.

### Allow multiple pack expansions in a tuple pattern

As currently proposed, we allow multiple `...` expressions within a tuple
literal expression, but only allow one `...` pattern within a tuple pattern. It
is superficially tempting to relax this restriction, but fundamentally
infeasible.

Allowing multiple `...` patterns would create a potential for ambiguity about
where their scrutinees begin and end. For example, given a signature like

> `fn F(... each xs: each i32, ... each ys: each i32)`, there is no way to tell

Reviewer 2: I think with latest versions, the `each` on `i32` isn't needed?

```suggestion
`fn F(... each xs: i32, ... each ys: i32)`, there is no way to tell
```

where `xs` ends and `ys` begins in the argument list; every choice is equally
valid. That ambiguity can be avoided if the types are different, but that would
make type _non_-equality a load-bearing part of the pattern. That's a very
unusual thing to need to reason about in the type system, so it's liable to be a
source of surprise and confusion for programmers, and in particular it looks
difficult if not impossible to usefully express with generic types, which would
greatly limit the usefulness of such a feature.

Function authors can straightforwardly work around this restriction by adding
delimiters. For example, the current design disallows
`fn F(... each xs: each i32, ... each ys: each i32)`, but it allows

> `fn F((... each xs: each i32), (... each ys: each i32))`, which is not only

Reviewer 2: Same as above:
```suggestion
`fn F(... each xs: i32, ... each ys: i32)`, but it allows
`fn F((... each xs: i32), (... each ys: i32))`, which is not only
```

unnecessary complexity to the code.

#### Keyword syntax

We could drop `...` altogether, and use a separate keyword for each kind of pack
expansion. For example, we could use `repeat` for variadic lists of tuple
elements, `do_repeat` for variadic statements, and `all_of` and `any_of` in
place of `...and` and `...or`. This leads to code like:

```carbon
// Takes an arbitrary number of vectors with arbitrary element types, and
// returns a vector of tuples where the i'th element of the vector is
// a tuple of the i'th elements of the input vectors.
fn Zip[repeat each ElementType:! type]
      (repeat each vector: Vector(each ElementType))
      -> Vector((repeat each ElementType)) {
  do_repeat var each iter: auto = each vector.Begin();
  var result: Vector((repeat each ElementType));
  while (all_of each iter != each vector.End()) {
    result.push_back((repeat each iter));
    repeat each iter++;
  }
  return result;
}
```

This approach has some major advantages: the keywords are more consistent with
`each` (and `expand` to some extent), substantially less visually noisy than
`...`, and they may also be more self-explanatory. However, it does have some

> substantial drawbacks.

Reviewer 2: It seems like we should mention & credit the Swift syntax explicitly here, which is similar but not exactly the same as this?

Author: Good point. How's this?

> # Variadics

Reviewer 3: Thinking of someone looking for this proposal in the future, isn't this little bit larger than just to be called the variadics?

Author: How so?

Reviewer 3: _pack_ and following things in the proposal look more general than variadics, and it occupies quite a lot.

Author: Huh. Packs aren't intended to be more general than variadics -- are you saying there are plausible non-variadic uses of them? If you're saying they aren't _necessary_ for variadics, then you might well be right (see the note on line 223), but I'm not sure yet, so I'm leaving them in the draft for now because I think I refer to them elsewhere in the doc.

Reviewer 3: It is fine to keep it variadic.

same arity. A pack expansion whose arguments have arity N is rewritten to N
copies of the pack template, where in the Kth copy, each use of the `[:]`
operator evaluates to the Kth element of its operand. Each copy is followed by a
separator, and the entire sequence is followed by a terminator, as determined by
the pack operator:

| Pack operator | Separator | Terminator |
| ------------- | --------- | ---------- |
| `...,`        | `,`       | (none)     |
| `...and`      | `and`     | `true`     |
| `...or`       | `or`      | `false`    |

As a consequence, a `...,` pack expansion can only occur where a tuple literal
element is expected, or directly inside parentheses, and in the latter case, the
parentheses form a tuple, even if the pack argument has only one element.

A pack expansion block has the form `...{` _statements_ `}`, and is rewritten to
N copies of the pack template `{` statements `}`, with pack arguments
substituted as above. `...{` is a single token, and may not contain internal
whitespace.

Pack expansions currently cannot be nested.

The spelling of `[:]` is taken from the C++ proposal
[P1858R2](http://wg21.link/P1858R2), but I'm very open to alternatives.

> Precedence of `[:]` is TBD, so I've erred on the side of over-parenthesizing.

Reviewer: FYI, you say it has the same precedence as prefix-`*` below.

### Examples

```carbon
// Computes the sum of its arguments, which are i64s
fn SumInts(..., [:]params: [i64;]) -> i64 {
  var sum: i64 = 0;
  for (var i: i64 in params) {
    sum += i;
  }
  return sum;
}
```

```carbon
// Concatenates its arguments, which are all convertible to String
fn StrCat[T:! [ConvertibleToString;]](..., [:]params: T) -> String {
  var len: i64 = 0;
  ...{ len += ([:]params).Length(); }
  var result: String = "";
  result.Reserve(len);
  ...{ result.Append(([:]params).ToString()); }
  return result;
}
```

```carbon
// Returns the minimum of its arguments, which must all have the same type T

> fn Min[template N:! BigInt where N > 0, T:! Comparable & Value]

Reviewer: FYI, We haven't specified how constraints on values will be written, but I don't think we are going to allow you to use `N` inside the type part of a binding for `N`.

Author: Hmm. I didn't intend it to be in the type part of the binding; I intended it to parse as `(template N:! BigInt) where N > 0`. In other words, it's a match guard, not a generic constraint operator (I think this example predated the latter).

I've changed this to `if` to avoid the ambiguity.

mostly reflects that earlier approach, and I'm not sure how much of it is still
relevant.

FIXME reassess if we still need pack values, once we have description of pattern
matching. Discussion of pack-type variables may be another reason we need it.
Even if we need them for that, we may not need them to be a separate category of
values; they could be tuple values, with different semantics imposed by the
type.

A _pack type_ is a kind of product type, like a tuple, and a value of a pack
type is called a _pack_. A pack type is itself a pack whose elements are types,
in the same way that a tuple type is itself a tuple whose elements are types. A
pack cannot be a sub-value of any other kind of value, although we intend to
support nested packs in the future. A name can be bound to a pack value, but
only if the name is local to a function.

Pack values are created using the `[:]` operator, which is a prefix unary
operator with the same precedence as unary `*`. If its operand is a tuple or
statically-sized array, its result is a pack containing the same elements, with
the same element types. For example, `[:]("a", "b", "c")` is a pack whose type

> is `[:](String, String, String)`, and whose elements are `"a"`, `"b"`, and

Reviewer: Technically this would be something closer to:
```suggestion
is `[:](LiteralString("a"), LiteralString("b"), LiteralString("b"))`, and whose
elements are `"a"`, `"b"`, and
```
This might be a concern in practice that literals don't necessarily have a common type.

operand of the `...,` pattern is iteratively matched with each of the remaining
K scrutinee elements, if any. The iterations are numbered sequentially from 1 to
K, and the number of the current iteration is called the _pack index_.

Observe that the pattern is irrefutable for a given scrutinee type if:

-   At most one `...,` pattern
-   Scrutinee type is a tuple type with at least N + M elements
-   First N and last M pattern elements are irrefutable given the corresponding
    scrutinee element types
-   For all I from 1 to K, the `...,` operand pattern with pack index I is
    irrefutable given the type of the Ith scrutinee element. (Note that the
    typechecker doesn't know K)

_Identifier that names a deduced parameter_

The deduced parameter is unified with the scrutinee value.

Pattern is irrefutable for a given scrutinee type if:

-   Unification does not reach a contradiction.

The typechecker must also ensure that the deduced value is a valid value of the
deduced parameter type.

_Name binding pattern:_

If the type subpattern contains a "free" use of the `[:]` operator, the name
binding pattern behaves at run-time as if it were prefixed with `[:]` (see

> below).

Reviewer: Example?

Author: I've put an example below, but I'm not sure how to usefully decompose it into examples for the individual rules, because they're pretty tightly coupled.

Reviewer: I am having trouble understanding this rule without an example, and I'm not sure which example below this rule applies to.

What does "free" mean here? Above you say `[:]` is only allowed inside a pack expansion, so it must mean something else, but I'm not sure what it would be.

What does "at run-time as if it were prefixed with `[:]`" mean? I thought the `[:]` were all rewritten away before we get to runtime.

Author: The example starting on line 404 contains an application of this rule: "That subpattern is a
name binding with a free `[:]` in the type subpattern, so we apply the rule for
name bindings prefixed with `[:]`". I don't know how to give a useful example of applying this rule in isolation, but I've added a brief example illustrating a case this rule applies to.

I've changed "free" to "unbound", and defined that term further up (see the second paragraph of "Expression and statement semantics").

I've dropped "at run-time"; I'm not sure why I put it there in the first place.

Author: I've rephrased this in terms of whether the name binding has a pack type, and explained more directly what happens in that case.

Otherwise, if the pattern is inside a pack expansion, but not inside a pack
argument pattern, the program is ill-formed (because the name would be bound

> multiple times).

Reviewer: Example?

Reviewer: What is unclear is how this is different from saying that you must use `[:]` somewhere in the pack expansion? I'm guessing the bad example is when the `[:]` isn't before the name, like `fn F(..., x: [:]___)` ? But that seems to be allowed in the `Zip` example above.

Author: I've added an example.

Name binding patterns are always irrefutable at run-time. However, typechecker
must also ensure the bound value is a valid value of the binding type.

> _`[:]` pattern:_

Reviewer: Example?

Reviewer: My issue here is that I don't think "`[:]` pattern" is on the list of kinds of patterns like "tuple pattern" and "binding pattern" are. Without an example of what you mean, I don't know what the subject of this section is.

Author: I've given an explicit definition, and added examples to the subcases.

I'm not sure what you mean by "on the list", but I'm basically defining these rules in terms of recursing down the AST, so I have more or less one case for each AST node kind that can appear in a pattern.

Reviewer: If this is about AST nodes, then perhaps each of these should start with the syntax rule it is applying to?

Author: I've simplified this section enough that I don't think I need headings in the first place. How's this?

FIXME maybe distribute these to the sections above?

Let I be the current pack index.

-   If the operand is a tuple pattern, it cannot contain a `...,` pattern, so
    all elements are explicit. The Ith element of the tuple pattern is matched
    with the scrutinee.
-   If the operand is an identifier that names a deduced parameter, the
    parameter must be bound to a K-tuple value (although it may have array or
    pack _type_), and its Ith element is unified with the scrutinee.
-   If the operand is a name binding pattern, the name must be bound to a
    K-tuple value, and its Ith element is bound to the scrutinee.
-   If the operand is an expression pattern, match the Ith element of its value

>     against the scrutinee.

Reviewer: Some examples would help here too

Author: OK, how's this?

substituted as above. `...{` is a single token, and may not contain internal
whitespace.

Pack expansions currently cannot be nested.

The spelling of `[:]` is taken from the C++ proposal
[P1858R2](http://wg21.link/P1858R2), but I'm very open to alternatives.

### Examples

```carbon
// Computes the sum of its arguments, which are i64s
fn SumInts(..., [:]params: [i64;]) -> i64 {
  var sum: i64 = 0;
  for (var i: i64 in params) {
    sum += i;
  }
  return sum;
}
```

```carbon
// Concatenates its arguments, which are all convertible to String
fn StrCat[T:! [ConvertibleToString;]](..., [:]params: T) -> String {
  var len: i64 = 0;
  ...{ len += ([:]params).Length(); }
  var result: String = "";
  result.Reserve(len);
  ...{ result.Append(([:]params).ToString()); }
  return result;
}
```

```carbon
// Returns the minimum of its arguments, which must all have the same type T
fn Min[template N:! BigInt if N > 0, T:! Comparable & Value]
    (..., [:]params: [T; N]) -> T {
  // Safe since params non-empty
  var result: T = params[0];
  for (var x: T in params.Slice(1..)) {
    if (x < result) {
      result = x;
    }
  }
  return result;

> }

Reviewer: I think this was originally my example, but I now think this may not be the way we want to express it in Carbon.

In particular, it would be nice if we could avoid all this reasoning about packs that we require to be non-empty by requiring the author of `Min` to write it like so:

```suggestion
// Returns the minimum of its arguments, which must all have the same type T
fn Min[template N:! BigInt, T:! Comparable & Value]
    (first: T, ..., [:]params: [T; N]) -> T {
  // Don't need to reason about N>0 to see this is safe
  var result: T = first;
  // Don't need to slice here
  for (var x: T in params) {
    if (x < result) {
      result = x;
    }
  }
  return result;
}
```

Author: Yeah, that's plausible. Separating out `first` might be awkward in some ways (easier for it to accidentally have a different type, for example), but at the moment I can't think of anything that seems like a dealbreaker.

Author: Done. I also dropped `N`, since we don't need it with this approach.

}
```

```carbon
fn Zip[ElementTypes:! [Type;]]
      (..., vectors: Vector([:]ElementTypes))
      -> Vector((..., [:]ElementTypes)) {
  var iters: auto = (..., vectors.Begin());
  var result: Vector((..., [:]ElementTypes));
  while (...and [:]iters != vectors.End()) {
    result.push_back((..., *[:]iters));
    ...{ ([:]iters)++; }
  }
  return result;
}
```

```carbon
// Toy example of mixing packs and values in a single parameter list.
// Takes an i64, any number of f64s, and then another i64.
fn MiddleVariadic(first: i64, ..., [:]middles: [f64;], last: i64);
```

## Details

FIXME throughout the following, clarify what parts apply to patterns.

### Statically sized arrays

`[T; N]` is the type of a statically-sized array of `N` objects of type `T`, and
`[T;]` is a pattern that matches the type `[T; N]` for any `N`. In other words,
it's a shorthand for `[T; _:! BigInt]`.

FIXME Or maybe we should treat it as an interface/constraint? See
https://discord.com/channels/655572317891461132/969001583088123905/973681495891910716
Or maybe it's `[T; template _:! BigInt]` so that we can validly refactor it to
`[template N:! BigInt]([T; N])`? `auto` seems like it might be ambivalent in a

> similar way, between `_:! Type` and `template _:! Type`.

Reviewer: This was up in the air for some time, but I believe the current thinking is we need `auto` to be special, in that it acts like a generic when matching something generic and like a template otherwise. In particular, we don't want `auto` to have the behavior of converting a generic to a template, like you might expect to happen if you pass the generic to a binding that explicitly includes the keyword `template`. #2153 is not resolved, though, so this is not final.

Author: #2153 is resolved now. Do I understand correctly that the resolution implies that `auto` does need to be special in this way?

Reviewer: `auto` does need to be special so that it matches the template/checked generic attribute of the argument. However, the resolution of #2153 is that both `_:! ...` and `template _:! ...` can match both template and generic arguments. So the choice is dependent on what you want to do with the size value, not what kind of value the caller has. In this case, what sort of type checking needs to be done with the size? The checked generic `_:! BigInt` is likely a better default unless the user specifically asks for a template.

Author: That makes sense. I'm going with that.

where in the Kth instance, every expansion argument is replaced by the Kth
element of that argument. The details of the rewrite vary slightly depending on
the root node of the expansion:

-   In a `...{` expansion, each instance uses a `{` in place of the opening
    `...{`.
-   In a `...,`, `...and`, or `...or` expansion, each instance has the expansion
    operator removed, and the instances are joined using the corresponding
    non-expansion operator.

An `...and` expansion with arity 0 is rewritten to `true` and an `...or`
expansion with arity 0 is rewritten to `false` (the identities of the respective
logical operations). A `...,` expansion with arity 0 is rewritten to `,`, and
any `,` immediately before or after the expansion is removed.

Even though an `[:]` operator can only appear within a pack expansion, it is
possible to apply these rewrite semantics to an isolated subexpression that
contains a `[:]` operator but not the enclosing pack operator. In those cases,
the result of the rewrite is an expression headed by a `[:]` operator, which
means that its value is a pack. Thus, we can say that
`([:]("a", "b", "c"), [:](1, 2, 3))` rewrites to

> `[:](("a", 1), ("b", 2), ("c", 3))`.

Reviewer: Is this needed for anything? Otherwise it seems like an interesting observation without a clear recipe for when it is applied, and I don't see what wouldn't still work if it were removed.

Author: At one point it was useful, at least as a specification convenience -- it lets us express the rewriting/evaluation logic in terms of recursion on the AST. I'm not sure if I'm still using it, but I'd like to keep it in the draft until we have a clearer idea of what's needed.

that's going to be an important property when typechecking variadic patterns.

_Tuple pattern:_

There can be no more than one `...,` pattern in a tuple pattern.

The N elements of the pattern before the `...,` expansion are matched with the
first N elements of the scrutinee, and the M elements of the pattern after the
`...,` expansion are matched with the last M elements of the scrutinee. If the
scrutinee does not have at least N + M elements, the pattern does not match. The
operand of the `...,` pattern is iteratively matched with each of the remaining
K scrutinee elements, if any. The iterations are numbered sequentially from 1 to
K, and the number of the current iteration is called the _pack index_.

Observe that the pattern is irrefutable for a given scrutinee type if:

-   At most one `...,` pattern
-   Scrutinee type is a tuple type with at least N + M elements
-   First N and last M pattern elements are irrefutable given the corresponding
    scrutinee element types
-   For all I from 1 to K, the `...,` operand pattern with pack index I is
    irrefutable given the type of the Ith scrutinee element. (Note that the

>     typechecker doesn't know K)

Reviewer: Do we allow anything where there is a single `...,` but it applies to multiple bindings each with a `[:]`
```
fn AddKeyValuePairsToContainer[addr self: Self*](
    ..., [:]keys: [KeyType;], [:]values: [ValueType;]);
// keys and values must have the same arity.
```
Not obvious whether it would mean `(key1, key2, key3, value1, value2, value3)` or `(key1, value1, key2, value2, key3, value3)`, but something like this does seem useful, and still well-defined as long as there are an even number of elements in the scrutinee.

Otherwise, looking at the examples above, I only see these cases:

-   `some_name: a_type(..., [:]some_array)` where `some_name` is going to bind to a single argument, but that argument has a type that varies in a variadic way, like `Apply`.
-   `..., [:]some_name: some_array_of_types`, like `SumInts`, `StrCat`, and `Min`
-   `..., some_name: a_type([:]some_array)`, like `Zip`

It seems better to just list these cases explicitly and say what they mean instead of trying to come up with more general rules.

Further, looking at these last two cases, I'm surprised by how different they look. Generalizing from the `Zip` example, I'd expect the `[:]` to be consistently to the right of the `:`, or to always have a `[:]` after `...,` before the `some_name:` in a binding.

Reviewer: To be clear, I'm not actually advocating for supporting that first example. While the `(key1, value1, key2, value2, key3, value3)` construct seems interesting, I'd be fine saying that isn't something we are going to support, and the caller will have to pass `((key1, value1), (key2, value2), (key3, value3))` instead. I do think it is a choice we should be making explicitly, though, either favoring expressiveness or reducing the surface area of this feature.

Author: Take a look at the `Zip` example, and in particular this line:

```
while (...and [:]iters != vectors.End())
```

That expression contains only one `[:]`, but in this context, `vectors` is a pack argument that's expanded implicitly (this is the crux of #1162), so this is an example of a pack expansion with two arguments.

I think you're right that we don't have any examples the arguments are both bindings (or even of doing this in a pattern rather than an expression or statement), but that seems more like an artifact of our limited set of examples, rather than something we can be confident will continue to be true with more use cases. I'm also not sure that disallowing this in patterns would actually make things any simpler.

But even if we do allow multiple packs in a pattern expansion, I don't think we would allow the specific example you gave. I would expect `..., [:]keys: [KeyType;], [:]values: [ValueType;]` to be parsed as two expressions separated by a comma, which means the second `[:]` isn't inside a pack expansion. As far as I can tell, in order for the user to fix this syntactic problem, they also have to fix the semantic ambiguity about whether they want `(key1, key2, key3, value1, value2, value3)`, `(key1, value1, key2, value2, key3, value3)`, or `((key1, value1), (key2, value2), (key3, value3))`.

(I'm not completely certain it's possible to write a pattern for `(key1, value1, key2, value2, key3, value3)`, and it's certainly more difficult than the other two, but that seems OK to me)

to appropriate sections of [`/docs/project/goals.md`](/docs/project/goals.md),
and/or to documents in [`/docs/project/principles`](/docs/project/principles).
For example:

-   [Community and culture](/docs/project/goals.md#community-and-culture)
-   [Language tools and ecosystem](/docs/project/goals.md#language-tools-and-ecosystem)
-   [Performance-critical software](/docs/project/goals.md#performance-critical-software)
-   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)
-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
-   [Practical safety and testing mechanisms](/docs/project/goals.md#practical-safety-and-testing-mechanisms)
-   [Fast and scalable development](/docs/project/goals.md#fast-and-scalable-development)
-   [Modern OS platforms, hardware architectures, and environments](/docs/project/goals.md#modern-os-platforms-hardware-architectures-and-environments)
-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)

> ## Alternatives considered

Reviewer: What was the reason to use prefix operators instead of postfix?

Reviewer: Of these, `[:]` is perhaps the most surprising, since having it postfix would be much more evocative of subscripting with `[`_`]`.

Author: Good point. Rather than answer in a comment, I've added my answer to the document.

## Table of contents

-   [Abstract](#abstract)
-   [Problem](#problem)
-   [Background](#background)
-   [Proposal](#proposal)
    -   [Examples](#examples)
-   [Details](#details)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Support variadic variable declaration statements](#support-variadic-variable-declaration-statements)
    -   [Support array expansion arguments](#support-array-expansion-arguments)
    -   [Omit variadic bindings](#omit-variadic-bindings)
        -   [Disallow pack-type bindings](#disallow-pack-type-bindings)
    -   [Fold expressions](#fold-expressions)
    -   [Allow multiple `...,` patterns in a tuple pattern](#allow-multiple--patterns-in-a-tuple-pattern)
    -   [Allow nested pack expansions](#allow-nested-pack-expansions)
    -   [Use postfix instead of prefix `...`](#use-postfix-instead-of-prefix-)
    -   [Require parentheses around `each`](#require-parentheses-around-each)
    -   [Keyword syntax](#keyword-syntax)

<!-- tocstop -->

## Abstract

Proposes a set of core features for declaring and implementing generic variadic
functions. The central concept is a _pack expansion_, which is a kind of

> compile-time loop over sequences called _expansion arguments_. Expansion

Reviewer 2: Terminology question:

What about calling the things that are expanded "packs" or "pack expressions" (instead of "expansion arguments")? That would match the term "pack expansion" better I feel like.

On the flip side, if we really like `expand` as the keyword, maybe we should change the outer terminology to something else? We could have "expansions" as the compile time loop over "expanded expressions"?

Author: I like your suggestion of "pack expansion" -> "expansion". After some consideration, I think "expansion argument" should be changed to "expansion parameter", but that seems like it might be orthogonal to your concerns. How would you feel about those two changes?

> What about calling the things that are expanded "packs" or "pack expressions" (instead of "expansion arguments")? That would match the term "pack expansion" better I feel like.

"Pack" already means something subtly different; I've revised to try to make that clear. "Pack expression" could maybe work, but I think it's liable to get confused with the importantly different concept of "expression that has a type pack". Also, they're not all expressions (a pattern can be an expansion argument).

> On the flip side, if we really like `expand` as the keyword, maybe we should change the outer terminology to something else? We could have "expansions" as the compile time loop over "expanded expressions"?

"Expanded expression" really sounds like it refers to `expand` in particular, whereas this term needs to also encompass usages of `each` bindings (and also _declarations_ of `each` bindings, which makes the "expression" part problematic).

arguments can be formed from tuples with the `expand` operator, or from a
_variadic binding_, which is marked with the `each` keyword at the point of
declaration and the point of use. Pack expansions come in several forms, with
slightly different iteration semantics:

-   `...,` iteratively evaluates an expression, and treats the values as

>     successive elements of an enclosing tuple.

Reviewer 2: Should this say something like:

```suggestion
-   `...,` iteratively evaluates an expression, and treats the values as
    successive elements of an enclosing tuple literal, struct literal, or argument list.
```

Struct literals seem important but maybe tricky, so might need to be deferred for now.

Author: "Tuple literal" is a good change, but "argument list" seems redundant with that, since argument lists are tuple literals. If you mean "parameter list", that's handled separately below; here I'm only talking about expressions, not patterns.

I don't really see why struct literals would be important (I'm having a hard time even coming up with plausible use cases), and they definitely look tricky, so I think we should defer that.

-   `...and` and `...or` iteratively evaluates a boolean expression, combining
    the values using `and` and `or`, and ending the loop early if the underlying
    operator short-circuits.
-   `...{}` iteratively executes a block of statements.

`...,` can also be used in patterns, where it iteratively matches the elements
of a scrutinee tuple. In conjunction with variadic bindings, this enables
functions to take an arbitrary number of arguments.

## Problem

Carbon needs a way to define functions and parameterized types that are
_variadic_, meaning they can take a variable number of arguments.

## Background

> C has long supported variadic functions (such as `printf`), but that mechanism

Reviewer 2: Maybe explicitly call these "varargs", which is the common term used to disambiguate them from C++ variadics?

Author: Done

several ways:

-   A type pack is composed of segments (like a tuple type), but is not a type.
-   A pack is a sequence of values (like a tuple), but is not a value.
-   An expression in a pack expansion does not have a type, but instead has a
    type pack.

We could resolve that awkwardness by instead saying:

-   A _pack type_ is composed of segments (like a tuple type).
-   A pack is a value, and its type is a pack type.
-   An expression in a pack expansion has a pack type.

However, that's confusing in a more fundamental way: it violates the basic
type-system property that the value of an expression is an instance of the type
of the expression.

It's tempting to try to resolve that problem by saying that the value of an
expression in a pack expansion is itself a pack. In effect, this treats pack
expansions like SIMD code, with expressions operating on "vectors" of data in
parallel, rather than iteratively executing the code on a series of scalar
values. And unfortunately, this model has the same limitation for variadics as
it does for SIMD: it can't model branching control flow.

For example, consider

> `(..., if (expand cond) then F(expand param) else G(expand param))`: if

Reviewer: ```suggestion
`(... if (expand cond) then F(expand param) else G(expand param))`: if
```

    type.

However, it also has some major disadvantages:

-   The implicit expansion of pack-type bindings hurts readability. For example,
    it's easy to overlook the fact that the loop condition
    `while (...and expand iters != vectors.End())` in `Zip` has two expansion
    arguments, not just one. This problem is especially acute in cases where a
    non-local name has a pack type.
-   We have to forbid template-dependent names from having pack types (see
    [leads issue #1162](https://github.com/carbon-language/carbon-lang/issues/1162)),
    because the possibility that an expression might be an expansion argument in
    some instantiations but not others would cause serious readability and
    implementability issues.
-   A given _use_ of such a binding really represents a single value at a time,
    in the same way that the iteration variable of a for-each loop does, so
    giving the binding a plural name and a pack type creates confusion in that
    context rather than alleviating it.
-   It may be confusing that a given expression might have either a pack type or
    a type pack, depending on whether it occurs in the type position of a
    binding pattern.

> -   It creates the possibility of

Reviewer: Incomplete sentence.

whereas `vector.Size()` could refer to the size of the sequence of bound values.
However, this could easily turn out to be a "wrong default": omitting `each`
seems easy to do by accident, and easy to misread during code review.

There are other solutions to this problem that work equally well with the status
quo or this alternative. In particular, it's already possible to express these
operations outside of a pack expansion by converting to a tuple, as in
`(... each vector).Size()` (status quo) or `(... vectors).Size()` (this
alternative). That may be sufficient to address those use cases, especially if
we relax the restrictions on nesting pack expansions. Failing that,
variadic-only spellings for these operations (like `sizeof...` in C++) would
also work with both approaches. So this issue does not seem like an important
differentiator between the two approaches.

#### Disallow pack-type bindings

As a variant of the above approach, it's possible to omit both variadic bindings
and pack-type bindings, and instead rely on variadic tuple-type bindings. For
example, the signature of `Zip` could instead be:

```carbon
fn Zip[ElementTypes:! [type;]]

>       (..., expand vectors: (..., Vector(expand ElementTypes)))
>       -> Vector((..., expand ElementTypes));

Reviewer: ```suggestion
      (... expand vectors: (... Vector(expand ElementTypes)))
      -> Vector((... expand ElementTypes));
```

difficult if not impossible to usefully express with generic types, which would
greatly limit the usefulness of such a feature.

Furthermore, our solution to the problem of combinatorial explosion in tuple
pattern matching deeply relies on the fact that there is at most one `...`
subpattern in the tuple pattern. Without that constraint, it's not at all clear
if we could type-check variadic pattern matches in polynomial time.

Finally, function authors can straightforwardly work around this restriction by
adding delimiters. For example, the current design disallows
`fn F(... each xs: each i32, ... each ys: each i32)`, but it allows
`fn F((... each xs: each i32), (... each ys: each i32))`, which is not only
easier to support, but makes the callsite safer and more readable, since the
boundary between the `xs` and `ys` arguments is explicitly marked. By contrast,
if we disallowed multiple `...` expressions in a function argument list,
function callers who ran into that restriction would often find it difficult or
impossible to work around. Note, however, that this workaround presupposes that
function signatures can have bindings below top-level, which is currently

> undecided.

Reviewer: Is there no alternative that doesn't involve destructuring function parameter patterns, where you write `fn F(xs: (... each i32), ys: (... each i32))` or something?

Author: I can see a few possible alternatives, but I don't love any of them:

- `F(xs: (... i32), ys: (... i32))`. See [this thread](https://github.com/carbon-language/carbon-lang/pull/2240#discussion_r1312315310) for further discussion.
- `F(xs: (... repeat i32), ys: (... repeat i32))`. I don't like introducing a new operator for such a specific use case, and I also can't find a good spelling for it: `repeat` seems like the most natural option based on its English meaning, but it could cause confusion because Swift variadics use the same keyword with a very different meaning.
- `F(xs: (... expand [i32;]), ys: (... expand [i32;]))`. This avoids the need for a new keyword, and it's a surprisingly small extension of concepts that we already have, but it's not exactly self-explanatory.

Allowing bindings below top level seems like the least bad option to me. In any event, we don't need to decide that yet.

it, rather than asking the compiler to infer it.

### Allow nested pack expansions

Earlier versions of this design allowed pack expansions to contain other pack
expansions. This is in some ways a natural generalization, but it added
nontrivial complexity to the design. In particular, when an expansion argument
is lexically within two or more pack expansions, we need a rule for determining
which pack expansion it's an argument to, in a way that is unsurprising and
supports the intended use cases. However, we have few if any motivating use
cases for it, which made it difficult to evaluate that aspect of the design.
Consequently, this proposal does not support nested pack expansions, although it
tries to avoid ruling them out as a future extension.

### Use postfix instead of prefix `...`

`...` is a postfix operator in C++, which aligns with the natural-language use
of "…", so it would be more consistent with both if `...`, `...and`, and `...or`
were postfix operators spelled `...`, `and...`, and `or...`, and likewise if
statement pack expansions were marked by a `...` at the end rather than the
beginning.

> However, prefixe syntaxes are usually easier to parse (particularly for humans),

Reviewer: ```suggestion
However, prefix syntaxes are usually easier to parse (particularly for humans),
```

leading and trailing elements are the types of the leading and trailing elements
of `L`, and whose expansion element is the type of the expansion element of `L`
(if any).

**Other expressions with variadic operands:**

Any other expression that has at least one operand with a variadic type is
typechecked as follows: first, we replace each variadic type with a
corresponding non-variadic representative type, and perform name resolution and
typechecking in terms of those types. Then, if the resulting type for the whole
expression contains any representative types, they are replaced with the
original types they represented. Finally, if any of the variadic types were
unbound, and the type of the whole expression would be a non-variadic type `T`,
it is instead given the type `[:][T;]`.

FIXME: That last step feels super ad-hoc, but without it, cases like
`(..., vectors.Size())` don't come out right. But _with_ it, cases like
`...{ len += vectors.Size(); }` don't come out right. Maybe we need to
distinguish expression from statement expansions? Maybe allow expansion with no
arguments once we've determined the (symbolic) arity? Maybe we give up on

> assigning a type to subexpressions inside an expansion? (Ick)

Reviewer: (I'm assuming `vectors` here is from the `Zip` example.)

Is the issue that you have to write `...{ len += ([:]vectors).Size(); }` instead? That only seems bad because `[:]` is a prefix operator.

What are the alternatives / degrees of freedom for this design decision?

Author: > (I'm assuming `vectors` here is from the `Zip` example.)

Yes, `vectors` is from the `Zip` example, so its type is `Vector([:]ElementTypes)` (and the type of `ElementTypes` is `[Type;]`). Also, `len` is intended to be an integer type, say `i32`.

> Is the issue that you have to write `...{ len += ([:]vectors).Size(); }` instead? That only seems bad because `[:]` is a prefix operator.

`vectors` is already a pack, so applying `[:]` to it wouldn't work. In any event, the original syntax (without the `[:]`) is valid and meaningful according to the run-time semantics we've specified -- it's equivalent to something like

```
{
  len += (..., vectors)[0].Size();
  len += (..., vectors)[1].Size();
  len += (..., vectors)[2].Size();
  // etc.
}
```

That being the case, the type-checking rules should accept it, but they don't:

Following the algorithm from the previous paragraph, we first typecheck `vectors.Size()` as if `vectors` had the representative type `Vector(A)`, where `A` is an archetype for `Type`, and come up with the type `i32` (or whatever `Size` returns) for the whole expression. But then that problematic final step kicks in and says that the type of `vectors.Size()` is `[:][i32;]`. That means that `len += vectors.Size()` won't typecheck, because `+=` expects both of its operands to be scalars.

If we omit that problematic final step, this example typechecks, but now `(..., vectors.Size())` doesn't typecheck: as before, the type of `vectors.Size()` is determined to be `i32`, but then that means we're applying the `...,` operator to an operand that doesn't have a variadic type, which isn't valid.

Now that I've said all that "out loud", I think maybe I have a solution (and I've even [talked about it before](https://discord.com/channels/655572317891461132/969001583088123905/1053087121726001193)): when we move into the domain of representative types, we need to keep track of "pack-ness" as a type-like property that's orthogonal to the type per se, kind of like value category in C++. So the type of `vectors` doesn't become `Vector(A)`, it becomes something like (`Vector(A)`, true). This property propagates outward through type checking, mostly unchanged, but it stops when it reaches the edge of a pack expansion. And actually, I think this property might need to be a number (or even symbolic value?) representing the pack arity, rather than just a boolean, so we can typecheck that arities are equal where appropriate.

That's probably not very clear; I'm mostly writing it down as a note-to-self. I'm happy to explain more if you like, and/or just write it up in the PR.

Author: I wound up going a somewhat different direction here, basically integrating these concerns directly into the type system rather than having a "pack-ness" side channel. In any event, I think this is fixed now.

    the values using `and` and `or`, and ending the loop early if the underlying
    operator short-circuits.
-   In a statement context, `...` iteratively executes a statement.
-   In a tuple literal pattern (such as a function parameter list), `...`
    iteratively matches the elements of the scrutinee tuple. In conjunction with
    variadic bindings, this enables functions to take an arbitrary number of
    arguments.

### Examples

```carbon
// Computes the sum of its arguments, which are i64s
fn SumInts(... each param: i64) -> i64 {
  var sum: i64 = 0;
  ... sum += each param;
  return sum;
}
```

```carbon
// Concatenates its arguments, which are all convertible to String
fn StrCat[... each T:! ConvertibleToString](... each param: each T) -> String {
  var len: i64 = 0;
  ... len += each param.Length();
  var result: String = "";
  result.Reserve(len);
  ... result.Append(each param.ToString());
  return result;
}
```

```carbon
// Returns the minimum of its arguments, which must all have the same type T.
//
// Note that this implementation is not recursive. We split the parameters into
// first and rest in order to forbid calling `Min` with no arguments.

> fn Min[T:! Comparable & Value](first: T, ... each next: T) -> T {

Reviewer 2: Perhaps this should be a follow-up, but when reading this it felt somewhat unfortunate.

Both that you needed to factor out the first element and then needed to explain why you factored it out because it is somewhat surprising.

I wonder if there is a reasonable syntactic way to represent 0-or-more vs. 1-or-more?

(This isn't a blocking comment, just didn't want to forget about it. If no one has an immediate idea, maybe just worth recording it in a possible future work section?)

Author: I agree that it's somewhat unfortunate, but as points in mitigation:

- The comment is speaking more to the reader of the _proposal_ (flagging a difference from C++), rather than the reader of the code in a real codebase.
- If we wrote the function signature without a separate `first`, the first line would need to be something like `var result: T = (... each arg)[0];`, which doesn't express the intent nearly so clearly.

I don't have any good ideas for a syntax (maybe somehow base it on the `+` vs `*` distinction in regexes?), but that's not the only problem: We would also need to extend the typesystem to model and propagate minimum-arity constraints on segments. That seems doable, and could be desirable for other reasons as well, but it seems substantial enough to belong in a follow-up proposal.

_variadic_, meaning they can take a variable number of arguments.

## Background

C has long supported variadic functions (such as `printf`), but that mechanism
is heavily disfavored in C++ because it isn't type-safe. Instead, C++ provides a
separate feature for defining variadic _templates_, which can be functions,
classes, or even variables. However, variadic templates currently suffer from
several shortcomings. Most notably:

-   They must be templates, which mean they typically must be defined in header
    files, are susceptible to code bloat due to template instantiation, and
    generally carry all the other costs associated with templating.
-   It is inordinately difficult to define a variadic function whose parameters
    have a fixed type, and the signature of such a function does not clearly
    communicate that fixed type to readers.
-   There is no procedural mechanism for iterating through a variadic parameter
    list. Although features like
    [fold expressions](https://en.cppreference.com/w/cpp/language/fold) cover
    some special cases, the only fully general way to express iteration is with

>     recursion, which is often awkward, and results in more template

Reviewer 4: ... and typically results in compile-time -- and sometimes runtime -- costs that grow quadratically in the length of the pack or worse.

representing the number of elements it evaluates to. All arguments of a given
pack expansion must have the same arity.

A pack expansion whose arguments have arity N is rewritten to N copies of the
pack expansion, where in the Kth copy, each expansion argument evaluates to the
Kth element of the pack. Each copy is followed by a separator, and the entire
sequence is followed by a terminator, as determined by the pack operator:

| Pack operator | Separator | Terminator |
| ------------- | --------- | ---------- |
| `...,`        | `,`       | (none)     |
| `...and`      | `and`     | `true`     |
| `...or`       | `or`      | `false`    |

As a consequence, a `...,` pack expansion can only occur where a tuple literal
element is expected, or directly inside parentheses, and in the latter case, the
parentheses form a tuple, even if the pack argument has only one element.

`...,` and `[:]` can also be used in patterns, with semantics chosen to follow
the general principle that pattern matching is the inverse of expression
evaluation.

> **Open question:** how should the pack argument operator be spelled? `[:]` is

> > taken from the C++ proposal [P1858R2](http://wg21.link/P1858R2), but there is

Reviewer 4: ```suggestion
> taken from the C++ proposal [P1858R2](http://wg21.link/P1858R2), but there it
```

> was motivated by a syntactic analogy with array indexing, which doesn't carry
> over to Carbon because it
> [needs to be a prefix operator](#use-postfix-instead-of-prefix-operators).
> `[:]` is also uncomfortably close to the `[T; N]` array type syntax, given
> that they will often be used in close proximity.

### Examples

```carbon
// Computes the sum of its arguments, which are i64s
fn SumInts(..., [:]params: [i64;]) -> i64 {
  var sum: i64 = 0;
  for (var i: i64 in params) {
    sum += i;
  }
  return sum;
}
```

```carbon
// Concatenates its arguments, which are all convertible to String

> fn StrCat[T:! [ConvertibleToString;]](..., [:]params: T) -> String {

Reviewer 4: I found this surprising: `T` is an array of types (more accurately, an array of facets), which is not a type, so I was not expecting it to be valid after a `:`. Are you envisioning a conversion from an array of types to a tuple type, analogous to the conversion from a tuple of types to a tuple type?

(I think I'd have found `..., params: [:]T` less surprising.)

Author: > Are you envisioning a conversion from an array of types to a tuple type, analogous to the conversion from a tuple of types to a tuple type?

Pretty much, yes. More precisely, I'm envisioning a conversion from an array to a tuple (and vice-versa), and I'm envisioning that we transitively close over all the tuple conversions. I've made that a little more explicit above, and elaborated more in `variadics.md`.

> (I think I'd have found `..., params: [:]T` less surprising.)

That would be valid too, but it would mean `params` has a pack type rather than a tuple type, which I think would somewhat hurt readability of the body in this case.

Note that even if Carbon supported more general C++-like fold expressions, we
would still probably have to give these three operators special-case treatment,
because `,` is a component of the tuple syntax rather than a freestanding binary
operator, and `and` and `or` are short-circuiting.

As a point of comparison, C++ gives special-case treatment to the same three
operators in fold expressions: they are the only ones where the initial value
can be omitted (such as `... && args` rather than `true && ... && args`) even if
the pack may be empty. Furthermore, folding over `&&` appears to have been the
original motivation for adding fold expressions to C++, and of course folding
over `,` was the central use case for C++ variadics.

Given that we are only supporting a minimal set of operators, allowing `...` to
occur in ordinary binary syntax has few advantages and several drawbacks:

-   It might conflict with a future general fold facility.
-   It would invite users to try other operators, and would probably give less
    clear errors if they do.
-   It would substantially complicate parsing and the AST.
-   It would force users to make a meaningless choice between `... or x` and

>     `... or y`, and likewise for `and`.

Reviewer 4: I don't follow this last bullet. The difference between the two options seems to be between `x` and `y`, which aren't names that are in scope here as far as I can tell.

Author: Whoops. Fixed.

readability and implementability that it is locally syntactically obvious
whether a given name refers to a pack, which rules out things like pack-type
data members.

Even with the existing restrictions, named packs may create readability
problems. Consider this example from earlier:

```carbon
fn Zip[ElementTypes:! [type;]]
      (..., vectors: Vector([:]ElementTypes))
      -> Vector((..., [:]ElementTypes)) {
  var iters: auto = (..., vectors.Begin());
  var result: Vector((..., [:]ElementTypes));
  while (...and [:]iters != vectors.End()) {
    result.push_back((..., *[:]iters));
    ...{ ([:]iters)++; }
  }
  return result;
}
```

It may not be obvious to the reader that the `...and` expression has two
expansion arguments, because only one of them is syntactically marked within the
expression.

One solution to these problems would be to disallow named packs, requiring user
code to convert packs to/from tuples when they need to be given names. For
example, with this approach, the signature of `Zip` would instead have to be:

```carbon
fn Zip[ElementTypes:! [Type;]]
      (..., [:]vectors: (..., Vector([:]ElementTypes)))
      -> Vector((..., [:]ElementTypes))

> ```

Reviewer 4: It looks like this contains a nested pack expansion, so I guess this alternative would also require that we revisit that choice.

Author: This code is actually valid under the current rules about nesting -- see the discussion in `variadics.md` (around line 95 at the moment), which actually uses this code as an example.

will often have those two operators in close conjunction, and readability is at
a premium in function signatures.

C solved a very similar problem with mixing prefix `*` and postfix `.` by
introducing a separate `->` operator. We could in principle solve this issue the
same way, introducing a separate "variadic binding operator" `$`, such that
`params$ [i64;]` is equivalent to `(params: [i64;])[:]`. However, this would
trade a readability problem for a teachability problem, by introducing more
syntax that programmers need to learn before they can even read variadic
function signatures.

### Keyword syntax

> **TODO:** Decide whether to make this the primary proposal, and restructure
> the discussion as appropriate.

We could build the syntax around keywords rather than symbolic tokens, much like
the
[Swift proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md).
For concreteness, the keywords I will discuss here are as follows:

| Symbolic token | Keyword replacement               |
| -------------- | --------------------------------- |
| `[:]`          | `expand`                          |

> | `...,`         | `repeat` and `unpack` (see below) |

Reviewer: I believe what was proposed in the discussion in person was `pack` instead of `unpack`

Author: Yeah, that's what we discussed, but on reflection it felt kind of confusing to me. Consider:

```
fn SumInts(pack params: expand [i64;])
```

`params` has a pack type, but the parameter list as a whole has a tuple type (as always), so if I read this like an expression, `pack` is an operation that consumes a pack. That feels backwards to me, because in ordinary usage, packing is the process of putting things into your pack, not the process of taking them out. I suppose you could read it differently, and say that the object of "pack" is the scrutinee, not the operand, so this means "take the scrutinee tuple, pack it, and then match it against the `params` binding pattern". I find that reading much less natural.

Based on your other comments it seems like you're envisioning a model where `pack` has different semantics than `...,`, as well as different syntax. The `pack` spelling seems more natural in that approach, where it reads as a noun rather than a verb.

Reviewer: > take the scrutinee tuple, pack it, and then match it against the `params` binding pattern

That was my reading, and seems natural in that a pattern is generally a description of how to match the scrutinee.

The different semantics of `pack` come from being used in a pattern context, where the arity is determined by the scrutinee rather than what follows the word `pack`.

Author: > > take the scrutinee tuple, pack it, and then match it against the `params` binding pattern
> 
> That was my reading, and seems natural in that a pattern is generally a description of how to match the scrutinee.

FWIW I don't read a pattern as a description of how to match the scrutinee, I read it as a description of the scrutinee itself. But the reason I find this reading unnatural is more fundamental: in English if I say "pack params", and I mean "pack" to be a verb, then "params" is the object of that verb, whereas you're saying the object of that verb is something else that isn't referred to anywhere in the utterance. Maybe a convention like that could be taught if patterns followed it consistently, but given that we almost never use verbs in patterns, I think that in the rare cases where we do, we need to avoid going against the grain of English usage as much as possible.

And to reiterate my previous point, I think your proposed semantic change substantially mitigates that problem, because it enables me to read `pack` as a noun, as in "pack _of_ params".

Reviewer 4: From discussion with @chandlerc: one issue that arises with pack parameters is whether the name is singular or plural. Singular works better for uses of the parameter, but (absent some suitable introductory keyword or similar) plural seems to work better when introducing the parameter. And even with the keyword `pack`, plural seems to work better than singular.

One option that we discussed and seems to potentially be promising is using `each` both when introducing the pack and when referencing the elements of it:

```
fn F[..., each T:! type](..., each x: each T) {
  ...{
    Print(each x);
  }
}
```

This both fixes the plurality problem and doesn't have the verb / noun problems that `pack` has.

Author: The usage of `T` in `each x: T` references the elements of the pack, so shouldn't it be `each x: each T`?

Reviewer 4: Yeah, it should, fixed.

> | `...,`         | `repeat` and `unpack` (see below) |

Reviewer 4: Have you considered using `...` instead of `...,`? While the consistency between `...,` and the other `...blah` tokens is nice, I think the presence of a `,` in the `...,` token is going to be a persistent readability issue, because it makes it visually unclear whether you're expanding the thing to the left or the thing to the right.

Eg:

```
fn F(...each x: i32) {
  if (...and each x > 0) {
    ...{
      Print(each x);
    }
    F(...each x - 1);
  }
}
```

In in-person discussion between @chandlerc, @josh11b and myself, there was some sentiment towards using prefix `...` in patterns and postfix `...` in expressions, but I'm not sure whether that will work well in the overall grammar, particularly when expressions appear within patterns.

I'm going to add some suggestions below using this syntax to make it easier to compare against other options.

Author: In your example, it seems like it will be really hard to resist reading `...each` as a folding operator, like `...and`. So to make this work, I think we'd have to require that the `...` token be followed by whitespace. With that amendment, this might be workable, although I do worry it could still be confusing.

Reviewer 4: I think reading `...each` as a folding operator actually works to our advantage in the pattern case, where there can't be any postfix pieces following the `each ...` subexpression. But it could lead to a false interpretation of the `F(...each x - 1)` case, and confusion about similar things like `F(...1 - each x)`. I'm not sure whether that would be a persistent confusion, or whether telling people to treat `each x` as the name of the variable would be sufficient to resolve the confusion permanently.

| `...and`       | `all_of`                          |
| `...or`        | `any_of`                          |

> | `...{`         | `repeat {`                        |

Reviewer: There are a lot of different keywords here, without a lot to connect them together. We considered: maybe all (or almost all) of these would be spelled using `...`: `expand...`, `repeat...`, `and...`, `or...`.

Not clear if `pack` should also be changed: it is different from the others in that it determines how many times to repeat from what it matches, and does not require any packs to be mentioned. The other pack keywords require some pack and use it to determine arity.

We tried some examples, though, and it seemed to be a bit much, particularly `repeat...` and `each...`.

Author: If we did something like that, I think I'd ideally want a consistent syntax pattern for all the expansion roots (`unpack`/`pack`, `repeat`, `all_of`, `any_of`), and a different pattern for the expansion arguments (`expand` and `each`). If `...` is the only tool in our toolkit, we should probably use it for the roots, because there are more of them, and because it seems more important to be able to quickly identify an expansion than to be able to quickly identify its arguments.

This approach seems to work best in combination with some other nominally
independent changes:

-   Forbid `expand` in pattern contexts.
-   Allow `repeat` only in expression contexts and `unpack` only in pattern
    contexts.
-   Require any usage of a pack-type variable to be prefixed with a new keyword

>     `each`.

Reviewer: @zygoloid says: this gives us syntactic room to give the pack itself an API, that can be accessed by referring to the pack without `each`. For example, a templated pack could support indexing (or maybe: indexing into a pack is a template operation):

```
fn ReverseTuple[template pack T:! type](x: (repeat each T))
    -> (repeat[i: i32] T[sizeof(T) - i - 1]) {
  return (repeat[i: i32] x[sizeof(T) - i - 1]);
}
```

Reviewer: (Expectation here is that `template` is needed to do pack indexing, since we do not expect to be able to type check these generically.)

We also discussed potentially getting the elements of the pack
```
fn F[template pack T:! type](x: (repeat each T)) {
    repeat[U in T, i: i32] {
        match (x) {
            // matches if every element of the tuple has the same type
            case (pack a: U) => { }
            // would match always
            case (pack a: each T) => { }
            // same as previous case
            case (repeat[V in T] a: V) => { }
        }
    }
}
```

Author: I agree that by requiring a syntactic marker to use a pack as an expansion argument, that creates syntactic space for using packs in other ways. However, I'm not so sure that's a good thing. If for every tuple type there's a corresponding pack type, and packs can be subscripted just like tuples, then a pack can do anything a tuple can do, and "should I use a pack or a tuple here?" starts to be a difficult question to answer. I can even imagine people wanting to be able to write functions that operate generically on either packs or tuples, so they don't have to write the same code twice to handle the two different families of product types. I'd much rather avoid opening that can of worms if at all possible.

I think your `ReverseTuple` example kind of supports that concern: it doesn't seem to rely on the fact that `T` is a pack in any meaningful way, and in fact it seems like it would be more straightforward if `T` were an array (i.e. a tuple type):

```
fn ReverseTuple[template T:! [type;]](x: T)
    -> ReverseArray(T) {
  return (repeat[i: i32] x[T.Length() - i - 1]);
}
```

Notice the simpler return type expression, and the fact that there are no packs anywhere in sight. 

As an aside, I don't think your hypothetical `repeat[]` construct will work as written -- the operands don't contain an `each` or `expand`, so there's no way to deduce the arity; it would have to be explicitly specified as a separate pseudo-argument, e.g. `repeat[i: i32; T.Length()]`. At which point we've basically invented a compile-time `for` loop, which I think is the feature this use case really wants; variadics are a red herring.

As for your second example, it seems to be fundamentally about working with nested expansions, which we've decided to treat as out of scope for now. However, I notice that in the last `case` you've used `repeat` in a pattern context, which is currently disallowed. Even if that was an accident, it's an interesting one, since we disallowed that because it was felt to be counterintuitive. If we can allow it, then we don't need `pack` in the first place.

With these changes, the examples given earlier would read as follows:

```carbon
// Computes the sum of its arguments, which are i64s

> fn SumInts(unpack params: expand [i64;]) -> i64 {

Reviewer: ```suggestion
fn SumInts(pack params: i64) -> i64 {
```

> // Computes the sum of its arguments, which are i64s
> fn SumInts(unpack params: expand [i64;]) -> i64 {
>   var sum: i64 = 0;
>   repeat {
>     sum += each params;
>   }
>   return sum;
> }

Reviewer 4: ```suggestion
// Computes the sum of its arguments, which are i64s
fn SumInts(...each param: i64) -> i64 {
  var sum: i64 = 0;
  ...{
    sum += each param;
  }
  return sum;
}
```

Reviewer 4: It's not clear to me that we need the braces in `...{...}`, in the case where we're expanding a single statement. I think we could handle:

```suggestion
// Computes the sum of its arguments, which are i64s
fn SumInts(...each param: i64) -> i64 {
  var sum: i64 = 0;
  ... sum += each param;
  return sum;
}
```

and potentially even

```carbon
fn Zip[...each ElementType:! type]
      (...each vector: Vector(each ElementType))
      -> Vector((...each ElementType)) {
  // Local pack.
  ...var each iter: auto = each vector.Begin();
  var result: Vector(...each ElementType);
  while (...and each iter != each vector.End()) {
    result.push_back((...*each iter));
    ...each iter++;
  }
  return result;
}
```

(There's an ambiguity here for a statement starting with `...{`, but it's easy to resolve with a single token of lookahead: `...{}` is invalid either way, `...{.` must be a single-statement expansion, and `...{` followed by anything else must be a block expansion.)

Author: Regarding omitting the braces, do you think it's important to resolve that now, or can we table it until the other syntax issues are settled?

Reviewer 4: I think we can table it. It's interesting that this syntactic approach has some potentially nice future generalizations, but I don't think they're motivating enough to be a primary factor when picking the syntax.

```

```carbon
// Concatenates its arguments, which are all convertible to String

> fn StrCat[T:! [ConvertibleToString;]](unpack params: expand T) -> String {

Reviewer: ```suggestion
fn StrCat[pack T:! ConvertibleToString](pack params: each T) -> String {
```

> // Concatenates its arguments, which are all convertible to String
> fn StrCat[T:! [ConvertibleToString;]](unpack params: expand T) -> String {
>   var len: i64 = 0;
>   repeat { len += (each params).Length(); }
>   var result: String = "";
>   result.Reserve(len);
>   repeat { result.Append((each params).ToString()); }
>   return result;
> }

Reviewer 4: ```suggestion
// Concatenates its arguments, which are all convertible to String
fn StrCat[...each T:! ConvertibleToString](...each param: each T) -> String {
  var len: i64 = 0;
  ...{ len += each param.Length(); }
  var result: String = "";
  result.Reserve(len);
  ...{ result.Append(each param.ToString()); }
  return result;
}
```

Author: Dropping the parentheses from e.g. `(each params).Length()` could be a problem. I expect we can pick a precedence to make that work from the parser's point of view, but as a human reader I find I really want to parse `each params.Length()` as `each (params.Length())`, because of the placement of the whitespace.

Reviewer 4: I think the mental model we'd want to teach for this `each` keyword is that `each param` should be treated as the name of the parameter. If people (and syntax highlighers) treat it that way, I think the parentheses in `(each param).Length()` might look as redundant as `(x).Length()` does. But I don't know that; it might be that visually splitting things up on whitespace is just too ingrained a habit for lots of readers. The misinterpretation is probably also not too harmful for readers of code, other than letting them infer an inaccurate rule -- reading `each param.Length()` as `each (param.Length())` or reading `each x + 1` as `each (x + 1)` still gives the right interpretation -- and for writers of code we should be able to give great diagnostics if the `each` is misplaced (eg, `each 1 - x` instead of `1 - each x`).

```

```carbon
// Returns the minimum of its arguments, which must all have the same type T.
//
// Note that this implementation is not recursive. We split the parameters into
// first and rest in order to forbid calling `Min` with no arguments.

> fn Min[T:! Comparable & Value](first: T, unpack rest: expand [T;]) -> T {

Reviewer: ```suggestion
fn Min[T:! Comparable & Value](first: T, pack rest: T) -> T {
```

> fn Min[T:! Comparable & Value](first: T, unpack rest: expand [T;]) -> T {
>   var result: T = first;
>   repeat {
>     if (each rest < result) {
>       result = each rest;
>     }
>   }
>   return result;
> }

Reviewer 4: ```suggestion
fn Min[T:! Comparable & Value](first: T, ...each next: T) -> T {
  var result: T = first;
  ...{
    if (each next < result) {
      result = each next;
    }
  }
  return result;
}
```

```

```carbon
// Invokes f, with the tuple `args` as its arguments.

> fn Apply[T:! [type;], F:! CallableWith(repeat expand T)]
>         (f: F, args: (repeat expand T)) -> auto {

Reviewer: ```suggestion
fn Apply[pack T:! type, F:! CallableWith(repeat each T)]
        (f: F, args: (repeat each T)) -> auto {
```

> fn Apply[T:! [type;], F:! CallableWith(repeat expand T)]
>         (f: F, args: (repeat expand T)) -> auto {
>   return f(repeat expand args);
> }

Reviewer 4: ```suggestion
fn Apply[...each T:! type, F:! CallableWith(...each T)]
        (f: F, args: (...each T)) -> auto {
  return f(...expand args);
}
```

or maybe, if we allow more interesting patterns in functions:

```suggestion
fn Apply[...each T:! type, F:! CallableWith(...each T)]
        (f: F, (...each arg: each T)) -> auto {
  return f(...each arg);
}
```

```

```carbon
// Takes an arbitrary number of vectors with arbitrary element types, and
// returns a vector of tuples where the i'th element of the vector is
// a tuple of the i'th elements of the input vectors.

> fn Zip[ElementTypes:! [type;]]
>       (unpack vectors: Vector(expand ElementTypes))
>       -> Vector((repeat expand ElementTypes)) {

Reviewer: ```suggestion
fn Zip[pack ElementTypes:! type]
      (pack vectors: Vector(each ElementTypes))
      -> Vector((repeat each ElementTypes)) {
```

Author: (I'm replying here on your [general suggestion](https://github.com/carbon-language/carbon-lang/pull/2240#pullrequestreview-1444521929) of using `pack` instead of deducing arrays of types, since it's hard to reply to top-level comments)

I don't think this follows from treating `pack` as a pattern-only alternate spelling of `...,` -- this is semantically as well as syntactically different from the primary proposal. In the primary proposal, `...,` takes a sequence of tuple elements as the scrutinee, converts them to a pack, and then matches that whole pack against the operand pattern. IIUC you're suggesting that instead it could match the _elements_ of the scrutinee pack against the operand pattern. In fact, it's really matching the elements against N copies of the operand pattern, which may be parameterized by the pack index.

That's not necessarily a problem, of course, I just want to make sure we agree that we're talking about different semantics, not just different syntax.

I do see the appeal of this approach; in particular, it's more concise, and it restores the declarative rather than imperative quality of the pattern grammar (although, ironically, it makes pattern semantics more imperative in the programming-language sense). At least on a first glance, it reads much better. However, I see a few potential problems:

- It means the type of a bound name is determined by both the binding pattern that declares it, and the context in which that binding pattern appears. In other words, the type is determined by the interaction of type information flowing both up and down through the AST. For example, in `StrCat`, the type of `T` is something like "pack of `ConvertibleToString`", where the `ConvertibleToString` part comes from a child node of the `T:!` binding, and the "pack" part comes from an ancestor node (usually the parent, but potentially much farther away).

  My guess is that this bidirectional information flow is sufficiently specialized that it won't lead to soundness problems in the type system. However, it does seem like a significant readability problem that if I want to know the type of a name, looking to the right of the `:` will give me the wrong answer.
- It means we have to use `pack` in deduced parameter lists. That's problematic because deduced parameter lists aren't really patterns, they just reuse pattern syntax. That means we'll need to specify the meaning of `pack` in a deduced parameter list, separately from what it means in a pattern. That will be annoying at the very least, because IIRC we haven't specified anything about deduction yet, so there's no foundation for the variadic case to build on. So far I've been able to handwave that with strategic use of the word "symbolic", but I doubt that will cut it if I need to support `pack` in deduced parameter lists.
- Arrays can sometimes be more convenient than packs. See my comment on how your `ReverseTuple` example can be rewritten to avoid using variadics at all, but only if we can deduce an array of types from a tuple type.

Reviewer: Calling @zygoloid , who was probably not notified of the updates to this thread.

* Re: "if I want to know the type of a name, looking to the right of the `:` will give me the wrong answer", the thing to the _left_ of the `:!` is `pack T` not `T`, so it seems reasonable that the thing to the right of the `:!` is not the type of `T`.
* Re: `pack` in the deduced parameter list -- the deduced parameter list is _part_ of the pattern if I'm not mistaken. It also has the "arity determined by the scrutinee" property.

Reviewer: Also, the intent of this suggestion is to be consistent with the behavior of expansions where non-packs values are repeated for each expansion.

Author: > * Re: "if I want to know the type of a name, looking to the right of the `:` will give me the wrong answer", the thing to the _left_ of the `:!` is `pack T` not `T`, so it seems reasonable that the thing to the right of the `:!` is not the type of `T`.

Oh, so is the intent that `pack` is part of the binding syntax (like `template`), so I can't say things like `pack (x: i32, y: i32)`? That would cost us some generality, but I don't know how much that matters -- evidently our motivating examples don't need it. In any event, that would mitigate my concern here, but it doesn't completely address it -- it still means that the type of a binding is _almost always_ the value of the expression in the type position, but in rare cases a keyword in a different position triggers a different rule. That seems like a confusing irregularity -- I'd much rather have something in the type position itself signal that the binding has a pack type.

> * Re: `pack` in the deduced parameter list -- the deduced parameter list is _part_ of the pattern if I'm not mistaken. It also has the "arity determined by the scrutinee" property.

The deduced parameter list is contained within a pattern, but that doesn't mean it itself is a pattern context. For example, the type position of a binding is contained within a pattern, but it's an expression context.

As examples of how the deduced parameter list seems different:
- Order is not significant. `T:! type, U:! i32` is equivalent to `U:! type, T:! i32` if it occurs in the deduced parameter list, and very much not equivalent if it occurs in a tuple pattern. 
- Names are significant. In terms of whether the pattern matches, `T:! type` is equivalent to `U:! type` if it occurs in any context _other than_ a deduced parameter list, but not equivalent in a deduced parameter list.
- Commas are structural. In a tuple pattern, `U:! type, pack T:! type` can be used to impose a structure that does not exist in the scrutinee -- "I will decompose this homogeneous list of types into a first and a rest". But in a deduced parameter list, it can't impose structure that didn't already exist in the scrutinee.

Speaking of which, if the deduced parameter list's contents are a pattern, what exactly _is_ the scrutinee? It seems like it will have to have some weird properties.

> Also, the intent of this suggestion is to be consistent with the behavior of expansions where non-packs values are repeated for each expansion.

Do you see this suggestion as being more consistent in that way than my version? If so, can you elaborate?

Reviewer 4: The more recent `each` suggestion recovers the generality here by separating the information that a binding is a pack from the position where it's expanded -- `(...(each x: i32, each y: i32))` would be a valid pattern, matching a list of pairs of `i32`s.

I think that approach also addresses your concern about the thing on the right of the `:` not necessarily being the type of the thing on the left -- when we declare `each x: T`, the type of `each x` is `T`. (@chandlerc described this as being somewhat like a name prefix / sigil, like `$` / `@` in perl.)

Regarding implicit parameter lists -- yeah, they're not really lists of general patterns, and they don't really have scrutinees (except for `self`). But I think it's important that moving a parameter between an implicit and explicit parameter list uses the same syntax. I think it would be surprising if you can directly introduce packs in pattern syntax, but you have to introduce an array when writing something analogous in an implicit parameter list. For example, I think it's valuable that these two functions use the same syntax for their packs of types:

```
// Make a tuple containing unformed values.
fn MakeUnformedTuple(...each T:! type) -> (...each T);

// Make a tuple containing specific values.
fn MakeTuple[...each T:! type](...each x: each T) -> (...each T);
```

Author: > I think that approach also addresses your concern about the thing on the right of the `:` not necessarily being the type of the thing on the left -- when we declare `each x: T`, the type of `each x` is `T`. (@chandlerc described this as being somewhat like a name prefix / sigil, like `$` / `@` in perl.)

I think this still has the same basic problem: we could no longer teach people that if you see `foo: Bar`, that means the type of `foo` is `Bar`; instead we have to teach them two rules, one for `foo: Bar` and one for `each foo: Bar`. In some ways it makes the problem worse: my initial concern was that "what is the type of `foo`?" has different answers depending on whether there's a `pack` in the surrounding context. But this new model seems to be saying that that question might not even have an answer, because sometimes we might only be able to talk about the type of `each foo`.

It also raises the question of whether Carbon has other cases where you can declare a name by giving the type of a composite expression involving it, rather than giving the type of the name itself. For example, maybe you can declare a pointer to `i32` as `*p: i32`? I assume we don't actually want to do that, but I think it will be a natural question for people learning the language to ask.

On the other hand, if we say that `foo` always has a type, that leads to odd irregularities: `each x: each T` has the same type as `x: each T`, but `each x: i32` has a different type than `x: i32`. So sometimes `each` is just an annotation, and sometimes it has a semantic effect.

> Regarding implicit parameter lists -- yeah, they're not really lists of general patterns, and they don't really have scrutinees (except for `self`). But I think it's important that moving a parameter between an implicit and explicit parameter list uses the same syntax. I think it would be surprising if you can directly introduce packs in pattern syntax, but you have to introduce an array when writing something analogous in an implicit parameter list. For example, I think it's valuable that these two functions use the same syntax for their packs of types:
> 
> ```
> // Make a tuple containing unformed values.
> fn MakeUnformedTuple(...each T:! type) -> (...each T);
> 
> // Make a tuple containing specific values.
> fn MakeTuple[...each T:! type](...each x: each T) -> (...each T);
> ```

I halfway agree: I think it's important (albeit perhaps not essential) that you can move a deduced parameter to the explicit parameter list without changing the syntax, but I don't think going in the opposite direction necessarily has to be so trivial. There are going to be all kinds of reasons why I might not be able to move an explicit parameter to the deduced parameter list, and the absence of pack expansions is going to be pretty far down the list. The primary proposal satisfies that more relaxed criterion, and consequently it allows you to write those functions using the same syntax:

```
fn MakeUnformedTuple(..., [:]T:! [type;]);
fn MakeTuple[T:! [type;]](..., [:]x: T);
```

The alternative described in this section doesn't quite meet that standard, because it doesn't have a pattern-context equivalent of `[:]`, and I do agree that's a drawback of this alternative. However, the change in syntax is still fairly minor:

```
fn MakeUnformedTuple(unpack T:! expand [type;]);
fn MakeTuple[T:! [type;]](unpack x: expand T);
```

  var iters: auto = (repeat (each vectors).Begin());

>   var result: Vector((repeat expand ElementTypes));

Reviewer: ```suggestion
  var result: Vector((repeat each ElementTypes));
```

  while (all_of expand iters != (each vectors).End()) {
    result.push_back((repeat *expand iters));
    repeat { (expand iters)++; }
  }
  return result;
}
```

```carbon
// Toy example of mixing packs and values in a single parameter list.
// Takes an i64, any number of f64s, and then another i64.

> fn MiddleVariadic(first: i64, unpack middles: expand [f64;], last: i64);

Reviewer: ```suggestion
fn MiddleVariadic(first: i64, pack middles: f64, last: i64);
```

```

```carbon
// Toy example of using the result of variadic type deduction.

> fn TupleConcat[T1s: [type;], T2s: [type;]]
>               (t1: (repeat expand T1s), t2: (repeat expand T2s))
>               -> (repeat expand T1s, repeat expand T2s) {

Reviewer: ```suggestion
fn TupleConcat[pack T1s: type, pack T2s: type]
              (t1: (repeat each T1s), t2: (repeat each T2s))
              -> (repeat each T1s, repeat each T2s) {
```

> taken from the C++ proposal [P1858R2](http://wg21.link/P1858R2), but there is
> was motivated by a syntactic analogy with array indexing, which doesn't carry
> over to Carbon because it
> [needs to be a prefix operator](#use-postfix-instead-of-prefix-operators).
> `[:]` is also uncomfortably close to the `[T; N]` array type syntax, given
> that they will often be used in close proximity.

### Examples

```carbon
// Computes the sum of its arguments, which are i64s
fn SumInts(..., [:]params: [i64;]) -> i64 {
  var sum: i64 = 0;
  for (var i: i64 in params) {
    sum += i;
  }
  return sum;
}
```

```carbon
// Concatenates its arguments, which are all convertible to String
fn StrCat[T:! [ConvertibleToString;]](..., [:]params: T) -> String {
  var len: i64 = 0;
  ...{ len += ([:]params).Length(); }
  var result: String = "";
  result.Reserve(len);
  ...{ result.Append(([:]params).ToString()); }
  return result;
}
```

```carbon
// Returns the minimum of its arguments, which must all have the same type T.
//
// Note that this implementation is not recursive. We split the parameters into
// first and rest in order to forbid calling `Min` with no arguments.

> fn Min[T:! Comparable & Value](first: T, ..., [:]rest: [T; N]) -> T {

Reviewer: ```suggestion
fn Min[T:! Comparable & Value](first: T, ..., [:]rest: [T;]) -> T {
```

    -   [Require parentheses around `each`](#require-parentheses-around-each)
    -   [Keyword syntax](#keyword-syntax)

<!-- tocstop -->

## Abstract

Proposes a set of core features for declaring and implementing generic variadic
functions. The central concept is a _pack expansion_, which is a kind of
compile-time loop over sequences called _expansion arguments_. Expansion
arguments can be formed from tuples with the `expand` operator, or from a
_variadic binding_, which is marked with the `each` keyword at the point of
declaration and the point of use. Pack expansions come in several forms, with
slightly different iteration semantics:

-   `...,` iteratively evaluates an expression, and treats the values as
    successive elements of an enclosing tuple.
-   `...and` and `...or` iteratively evaluates a boolean expression, combining
    the values using `and` and `or`, and ending the loop early if the underlying
    operator short-circuits.
-   `...{}` iteratively executes a block of statements.

`...,` can also be used in patterns, where it iteratively matches the elements
of a scrutinee tuple. In conjunction with variadic bindings, this enables

> functions to take an arbitrary number of arguments.

Reviewer: This is great. Perhaps copy it as an intro into `variadics.md`?

Author: I worry that as an introduction it would put off a lot of readers, because it's so compressed and has no examples. Am I thinking about that the wrong way?

Reviewer: I think it is useful as a summary of what you are going to say, giving a high level view. My expectation is that this summary will give the reader a fuzzy global view where they don't really understand everything, but they can then refine by reading further. Right now `variadics.md` is very hard to understand since it goes into depth into syntax and terminology before talking at all about semantics. So the reader has no framework for understanding the syntax when it is being explained.

I think an even better structure is:
- This summary
- The simplest example, `SumInts`, end to end. This should include the explanation of each new syntax being introduced, one at a time, and the meaning of the result given an example.
- Another example that adds one feature (`Min`), to demonstrate that feature, along with an explanation of that feature.
- Repeat until all features have been explained.

That is a big change, though, and I think it would be reasonable to just move the variadics.md document wholesale into the details section of the proposal and delete the variadics.md file from this PR. Once the proposal is approved, a follow-up PR can iterate on creating variadics.md.

Author: > I think it is useful as a summary of what you are going to say, giving a high level view. My expectation is that this summary will give the reader a fuzzy global view where they don't really understand everything, but they can then refine by reading further. Right now `variadics.md` is very hard to understand since it goes into depth into syntax and terminology before talking at all about semantics. So the reader has no framework for understanding the syntax when it is being explained.

I see, that makes a lot of sense. Done.

> I think an even better structure is:
> 
> * This summary
> * The simplest example, `SumInts`, end to end. This should include the explanation of each new syntax being introduced, one at a time, and the meaning of the result given an example.
> * Another example that adds one feature (`Min`), to demonstrate that feature, along with an explanation of that feature.
> * Repeat until all features have been explained.
> 
> That is a big change, though, and I think it would be reasonable to just move the variadics.md document wholesale into the details section of the proposal and delete the variadics.md file from this PR. Once the proposal is approved, a follow-up PR can iterate on creating variadics.md.

Thanks, I really like that approach. I might try to do that in this PR, because the proposal can't move forward if people don't understand it.

Reviewer 2: Just following up here with a repeat of what I think came up in the open discussion today --

I think actually for the proposal, we've got enough context to understand it and let it move forward with the simple suggestion from Josh and for the larger change to be a follow-up to make the design docs work more broadly. We've been discussing this enough that at least for me, it's gelling nicely, and I'm inclined to not lose that momentum for a more complex change to the document. =]

        -   [Fold-like syntax](#fold-like-syntax)
        -   [Variadic blocks](#variadic-blocks)
        -   [Keyword syntax](#keyword-syntax)
    -   [Require parentheses around `each`](#require-parentheses-around-each)
    -   [Fused expansion tokens](#fused-expansion-tokens)

<!-- tocstop -->

## Abstract

> **TODO:** change "pack expansion" to "expansion", and change "expansion
> argument" to "expansion parameter".

Proposes a set of core features for declaring and implementing generic variadic
functions.

A "pack expansion" is a syntactic unit beginning with `...`, which is a kind of
compile-time loop over sequences called "packs", which are specified by
"expansion arguments". Expansion arguments can be formed from tuples with the
`expand` operator, or from a "variadic binding", which is marked with the `each`
keyword at the point of declaration and the point of use.

The syntax and behavior of a pack expansion depends on its context, and in some
cases by a keyword following the `...`:

-   In a tuple literal expression, `...` iteratively evaluates its operand
    expression, and treats the values as successive elements of the tuple.

> -   `... and` and `... or` iteratively evaluate a boolean expression, combining

Reviewer: Didn't you suggest the preferred form of these is without the space?

```suggestion
-   `...and` and `...or` iteratively evaluate a boolean expression, combining
```

## Proposal

`[T; N]` is the type of a statically-sized array of N objects of type `T`.
`[T;]` is a shorthand for the type of an array of objects of type `T`, whose
size is static but deduced.

`...,`, `...and`, and `...or` are prefix unary expression operators with the
same precedence as the corresponding non-pack operators. `...{` is an opening
delimiter that forms a balanced pair with `}`, and can contain any number of

> `;`-delimited statements. An AST rooted at any of these operations is called a

Reviewer: My understanding is that statements include their terminating symbol, and sometimes it is `}` not `;`.
```suggestion
delimiter that forms a balanced pair with `}`, and can contain any number of
statements. An AST rooted at any of these operations is called a
```

Swift has very recently adopted a
[proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md)
to support definition-checked generic variadics. There have been several
attempts to add such a feature to Rust, but that work is
[currently inactive](https://github.com/rust-lang/rfcs/issues/376#issuecomment-830034029).

> ## Proposal

Reviewer: I feel like this part of the proposal and variadics.md necessarily overlap a lot. Would it make sense to combine Proposal and Examples sections into variadics.md and just put a pointer to that from here, like you do for the details section? (or only have the proposal for now and worry about extracting the text into variadics.md as a separate step)? For instance, the examples here would be very helpful context to include in variadics.md.

Author: I agree that the redundancy is awkward, but replacing this stuff with a pointer to variadics.md seems potentially awkward too, since variadics.md might change in the future. I'm especially hesitant to remove the examples from this document, because they're cited repeatedly (and changing them in variadics.md wouldn't even require a proposal).

I've copied the examples into variadics.md.

Reviewer: (as suggested in my other comment:) I think this would be solved by moving all the content into p2240.md and deleting variadics.md. That would give much more context to the content in variadics.md and avoid having to duplicate anything or have pointers.

Author: You're right that the way I have it now involves a fair amount of redundancy and means the proposal document isn't self-contained. However, moving the content of `variadics.md` into the proposal would mean that we risk going for a long time with design docs that are out of sync with the adopted proposals. You could argue that it won't be a long time because I can just copy the content from `p2240.md` back to `variadics.md` in a separate PR, but if so, I think to a large extent that's because I've been "keeping myself honest" by keeping `variadics.md` separate up until now. Based on your comment that merging the two would "give much more context to the content in variadics.md", it sounds like even with the two docs separated, `variadics.md` doesn't successfully stand on its own -- imagine how much worse that would have been if I wasn't even intending it to be a stand-alone document.

Also, the approach of putting everything in the proposal doc doesn't generalize well to proposals that modify existing docs rather than introducing wholly new ones.

(I'm incorporating a reply to [this comment](https://github.com/carbon-language/carbon-lang/pull/2240#pullrequestreview-1623559901) here because top-level PR comments don't thread:)

> I still really think that `variadics.md` should be inlined into `p2240.md`, but that can wait until people are done reviewing to avoid churn.

That seems like the worst of both worlds: PR reviewers still have to jump between documents, I still have to keep the duplicate passages in sync, future readers don't see the proposal that was actually reviewed (and have to deal with whatever awkward seams result from the last-minute inlining), and yet we don't wind up with a discoverable design doc for variadics after the proposal merges.

Reviewer: We as a project have so far decided to allow proposals without corresponding design doc changes. I think trying to split the content into the proposal and `variadics.md` isn't helping here:

* `variadics.md` right now does not stand well on its own, and is very different from the form we anticipate it having. Attempts to make `variadics.md` stand on its own have introduced a bunch of redundancy.
* The proposal text is not complete, and instead has a pointer to `variadics.md` for important content. This is going to make it harder to understand what proposal was actually agreed to, since `variadics.md` is going to change significantly.

If you think it is better to merge them now rather than wait, I think that would be totally fine, I was just trying to give you another option if you were worried about in-flight comment threads.

## Proposal

A "pack expansion" is a syntactic unit beginning with `...`, which is a kind of
compile-time loop over sequences called "packs". The packs that an expansion
iterates over are generated by syntactic units called "expansion arguments".
Expansion arguments can be formed from tuples with the `expand` operator, or
from a "variadic binding", which is marked with the `each` keyword at the point
of declaration and the point of use.

The syntax and behavior of a pack expansion depends on its context, and in some
cases by a keyword following the `...`:

> -   In a tuple literal expression, `...` iteratively evaluates its operand

Reviewer: Maybe it is worth stating explicitly that this includes function call argument lists?

    expression, and treats the values as successive elements of the tuple.
-   `... and` and `... or` iteratively evaluate a boolean expression, combining
    the values using `and` and `or`, and ending the loop early if the underlying
    operator short-circuits.
-   In a statement context, `...` iteratively executes a statement.

> -   In a tuple literal pattern, `...` iteratively matches the elements of the

Reviewer: Maybe worth stating that this includes function declaration parameter lists?

element is expected, or directly inside parentheses, and in the latter case, the
parentheses form a tuple, even if the pack argument has only one element.

> Pack expansion syntax can also be used in patterns, with semantics chosen to

Reviewer: "Pack expansion syntax here" means `...,`? Does it also include `[:]`? What about `...and` / `...or`? I guess it is `...,` followed by expansion arguments (which may include `[:]`)? It is a little unclear since the term is being used without first being defined.

Author: How's this?

```carbon
// Concatenates its arguments, which are all convertible to String

> fn StrCat(... each param: some ConvertibleToString) -> String {

Reviewer: Why this change? It seems like it would not be desirable to use `some` which is not in any accepted proposal.

Author: Wow, I have no idea where that came from. Fixed.

//
// Note that this implementation is not recursive. We split the parameters into
// first and rest in order to forbid calling `Min` with no arguments.
fn Min[T:! Comparable & Value](first: T, ... each next: T) -> T {
  var result: T = first;
  ... if (each next < result) {
    result = each next;
  }
  return result;
}
```

```carbon
// Invokes f, with the tuple `args` as its arguments.

> fn Apply[... each T:! type, F:! CallableWith(... each T)](f: F, args: (... each T)) -> auto {

Reviewer: Wrapping inside ``` ... ``` blocks is manual, and the result is ugly if the contents are too wide.
```suggestion
fn Apply[... each T:! type, F:! CallableWith(... each T)]
    (f: F, args: (... each T)) -> auto {
```

```carbon
// Toy example of using the result of variadic type deduction.
fn TupleConcat[T1s: [type;], T2s: [type;]](
    t1: (..., [:]T1s), t2: (..., [:]T2s)) -> (..., [:]T1s, ..., [:]T2s) {
  return (..., [:]t1, ..., [:]t2);
}
```

## Details

### Statically sized arrays

`[T; N]` is the type of a statically-sized array of `N` objects of type `T`, and
`[T;]` is a pattern that matches the type `[T; N]` for any `N`. In other words,
it's a shorthand for `[T; _:! BigInt]`. Consequently, for purposes of symbolic
evaluation and typechecking, we will assume that all statically-sized array
types have a size, although that size may be a symbolic value rather than a
constant.

An array can be initialized from a tuple, so long as the tuple has the right
number of elements, and all element types are convertible to the array element
type. We do not support deducing the element type of an array, as in
`fn F[T:! type](array: [T;]);`, for the same reason that we do not support

> `fn F[T:! type](x: T, y: T)`. This limitation may be addressed in future work.

Reviewer: I think the situation is different in two important ways:

- We do support `fn F[T:! type](x: T, y: T)`. What we don't support is calling `F` with two different types, and automatically figuring out a common type to implicitly convert to (like we do with `fn F(x: i32, y: i32)`). This is a limitation that may be addressed in future work.
- We can't support `fn F[T:! type](array: [T;]);` since it is legal to call `F()` with no arguments, and therefore no way to determine `T`. This does not seem like something that could be addressed in future work.

Author: Good point. I think that second limitation could be addressed, just not in a way that helps for variadics -- we could say the caller has to cast if the array has size 0, e.g. `F(() as [i32;])`. I've rephrased to try to thread that needle without taking up too much space.

fn MiddleVariadic(first: i64, ..., [:]middles: [f64;], last: i64);
```

```carbon
// Toy example of using the result of variadic type deduction.

> fn TupleConcat[T1s: [Type;], T2s: [Type;]](

Reviewer: FYI, `Type` is spelled `type` as of #2360 

```suggestion
fn TupleConcat[T1s: [type;], T2s: [type;]](
```

Author: Yeah, I'm going to make a pass through the whole doc for that.

The rest of the design for statically-sized arrays is deferred to a separate
proposal, because it's not relevant to variadics.

### Syntax and semantics of variadics

`...,`, `...and`, and `...or` are prefix unary expression operators with the
same precedence as the corresponding non-pack operators. They are single tokens,
and may not contain internal whitespace. `...{` is an opening delimiter that
forms a balanced pair with `}`, and can contain any number of `;`-delimited
statements. An AST rooted at any of these operations is called a _pack
expansion_.

A pack expansion must contain one or more _expansion arguments_. These are
usually marked by the `[:]` operator, but can also be unmarked usages of names
that have pack types (which will be defined below). The _arity_ of an expansion
argument is a compile-time value representing the number of elements it
evaluates to. Every pack expansion must contain at least one expansion argument,
and all arguments of a given expansion must have the same arity (which we will
also refer to as the arity of the expansion). If any expansion argument is a
pattern, the entire pack expansion is a pattern. In particular, this means that

> the expansion arguments of a `...{` expansion must be expressions.

Reviewer: This seems to preclude things I think we'd want to allow, like:
```
fn SumOfNorms[ElementTypes:! [Number;]]
      (..., vectors: Vector([:]ElementTypes))
      -> f64 {
  var iters: auto = (..., vectors.Begin());
  var result: f64 = 0.0;
  while (...and [:]iters != vectors.End()) {
    ...{
      let x: [:]ElementTypes = *[:]iters;
      result += (x * x).AsFloat();
      ([:]iters)++;
    }
    return result;
  }
```
Or is that using an expansion argument in a pattern?

Author: Your concern being the `let x: [:]ElementTypes = *[:]iters;`? I think that's allowed, because `ElementTypes` is an expression, and it's being used in an expression context (the lhs of a binding is an expression). I think we should even allow arguments of a `...{` to be used in a pattern context, so long as they are expression patterns, and I've tweaked the phrasing to be clearer about that.

That said, this _would_ disallow `let [:]x: ElementTypes = *[:]iters;`, which seems like a reasonable thing to want to write. I think I may see how we can drop this restriction (see also the new TODO), but I need to think about it more.

Pack expansions can be nested only if the inner expansion is within one of the
outer expansion's arguments, although we may relax this restriction in the

> future.

Reviewer: This would benefit from an example. Or is there one of the examples above that does this?

Author: Added an example.

evaluation, but does not naturally extend to statement semantics.

We will rely primarily on the reified model, with two exceptions:

-   We will use the syntactic model to describe the run-time semantics of `...{`
    expansions, because the reified model can't easily describe the semantics of
    statements, and it's difficult to bridge between the two models within a
    single pack expansion.
-   We will sometimes use the syntactic model when describing the run-time
    semantics of expressions, because expressions inside `...{` expansions must
    use that model, and because run-time expressions in other expansions will
    likely be implemented using that model.

Type-checking and pattern-matching will always use the reified model, because
they are inherently symbolic.

### Syntactic semantics

In the syntactic model, the semantics of packs and pack expansions are specified
in terms of a procedure for rewriting an AST rooted at `...{`, `...,`, `...and`,
or `...or` to an equivalent AST that does not. This rewrite takes place at the
same time as monomorphization of generic functions, which means it takes place

> after name resolution and typechecking.

Reviewer: It happens after checked generic typechecking, but not as clearly after template typechecking.

It also takes place when the values of symbolic constants are known, which is important to the following.

Author: How's this?

Under some circumstances, a variadic expression must be evaluated symbolically,
because the values of one or more of its expansion arguments is not yet known.
For example, this is often needed when variadic operations are used in a type
position. To support that, we need to introduce some new kinds of values to
represent the inputs and outputs of indefinite variadic operations.

A _pack element_ represents a sequence of an indefinite number of values that
all match a single pattern. The pattern and the size of the sequence are
represented as
[symbolic values](/docs/design/README.md#value-categories-and-value-phases) (or
constants). (FIXME: The term "pattern" can be kind of confusing in actual usage.
It's also arguably a category error: these things relate to patterns in roughly
the same way that values relate to expressions.) The symbolic value representing
the pattern may use special variables called _representatives_, which represent
the components of the pattern that can vary across values in the sequence.
Consequently, the scope of a representative is the pack element it appears in,
rather than some lexical scope. The size of a pack is represented as a sum of

> positive integer constants and symbolic variables of type `BigInt`. As their

Reviewer: Maybe say that these symbolic variables are known to be non-negative?

name suggests, pack elements cannot occur on their own, but only as an element
of an enclosing pack.

A _pack_ is a value consisting of a sequence of pack elements. Pack values are
always assumed to be normalized, meaning that every element's size is either a
symbolic value consisting of a single variable name, or a constant 1. This is

> always possible because a pack element size is always a sum of positive values,

Reviewer: Sum of non-negative values? Don't we support packs of length 0?

Author: Well, if an element is known to have size 0 it should be deleted during normalization if not earlier, but yes, the symbolic variables include 0 in their range. How's this?

so an element that doesn't fit this criterion can be split into a sequence of
elements with the same pattern that do. The _shape_ of a pack is another pack,
which is found by replacing each element's pattern with `()`. A _pack type_ is
the type of a pack, and can be represented as a pack value whose element
patterns match types. The type of a pack is found by replacing each of its
element's patterns with the type of that element's pattern. Thus, the
relationship between packs and pack types is much the same as between tuples and
tuple types.

During symbolic evaluation, a tuple value consists of a sequence of pack
elements, just like a pack. In other words, tuple types and pack types have the
same sets of values, and differ only in the operations they support. Name

> bindings can have pack types, but only if they are local to a single function.

Reviewer: Can you give an example of name bindings with pack types? It seems like there are two cases: inside a `...{`...`}` and outside.

Author: Sure, added an example.

I've also added text elsewhere to specify that if an expansion argument is a pattern, the whole expansion must be a pattern, and as a corollary, the expansion arguments of `...{}` must be expressions. I imposed that restriction because `...{}` is specified using the syntactic model, whereas pattern matching is specified using the reified model. I don't see a way to move either of them to the other model, or to bridge between the two models within a single expansion.

-   If the element is headed by `...,`, we evaluate its operand to produce a
    pack, and append its elements to the result tuple.
-   Otherwise, we evaluate the element expression to produce a value `V`, and
    append a pack element with pattern `V` and size 1 to the result tuple.

The `[:]` operator transforms a tuple into a pack, or an unknown tuple value
into a pack of unknown values:

-   When the operand is a tuple value, the result is a pack value that's
    identical to the tuple value.
-   When the operand is a symbolic variable whose type is a tuple, the result is
    a pack with the same shape as the tuple. Each element's pattern is a
    representative whose type is the pattern of the corresponding element of the
    tuple.

For these purposes, an array type `[T; N]` is treated as a tuple type
`(..., [:][T; N])`, so:

-   When the operand is an array type `[T; N]`, the result is a pack consisting
    of a single pack element `..., [:][T; N]`
-   When the operand is a symbolic variable with type `[T; N]`, the result is a
    pack consisting of a single pack element whose pattern is a representative
    whose type is `T`.

> An identifier expression that names a variable of pack type behaves like a `[:]`

Reviewer: Perhaps clarify that the type of the variable is pack type, not the value of the variable?

</td>
<td>

```cpp
template <typename T, typename... Params>
    requires std::totally_ordered<T> && std::copyable<T> &&
             (std::same_as<T, Params> && ...)
T Min(T first, Params... rest) {
  if constexpr (sizeof...(rest) == 0) {
    // Base case.
    return first;

>   else {

Reviewer: ```suggestion
  } else {
```

logical operations). A `...,` expansion with arity 0 is rewritten to `,`, and
any `,` immediately before or after the expansion is removed.

### Reified semantics and pack values

Unlike syntactic semantics, the reified semantic model is used at compile time,
when values may be unknown, or contain unknowns. Crucially, that can include
pack arities, so our representation of packs must be able to accommodate that.

A _pack element_ represents a sequence of some definite or indefinite number of

> values that all match a single value, which is called the _kernel_. During

Reviewer: I'm having trouble with this sentence:

- Is it framed like the definition of a pack element, but instead seems to tell you what they do.
- Part of the confusion here seems to be that this is explicitly in the context where the elements of a pack are unknown?
- This is talking about values matching values? I thought patterns match values.
- In general, some context seems to be missing.
- Unclear what the difference between a pack element and a kernel is here.

Maybe it could be explained in the context of an example?

Author: I've overhauled this section based on our live discussion, but I've opted for a series of more focused examples rather than one big one.

symbolic evaluation, the kernel and the arity of the sequence are represented as
[symbolic values](/docs/design/README.md#value-categories-and-value-phases) (or
constants). The symbolic value representing the kernel may use special variables
called _representatives_, which represent the components of the kernel that can
vary across values in the sequence. Consequently, the scope of a representative
is the pack element it appears in, rather than some lexical scope. The arity of
a pack is represented as a sum of positive integer constants and symbolic
variables that are known to be non-negative. At run time, the kernel will a
non-symbolic value, and the arity will always be a constant. As their name
suggests, pack elements cannot occur on their own, but only as an element of an
enclosing pack.

> **TODO:** find a better term than "kernel", if possible.

> A _pack_ is a value consisting of a sequence of pack elements. Pack values are

Reviewer: Why is this definition equivalent to the definition used above?

> The value of each expansion argument is a _pack_, which is a sequence of values much like a tuple.

Author: Is this better now?

    result is a pack with the same shape as the tuple. Each pack component's
    representative is `X[$I]`, and its type is the representative of the
    corresponding component of the tuple.

For these purposes, an array type `[T; N]` is treated as a tuple type
`(..., [:][T; N])`, so:

-   When the operand is an array type `[T; N]`, the result is a pack consisting
    of a single component with representative `T` and arity `N`.
-   When the operand is a symbolic variable `X` with type `[T; N]`, the result
    is a pack consisting of a single pack component whose representative is
    `X[$I]`, and whose arity is `N`.

An identifier expression that names a variable whose type is a pack type behaves
like a `[:]` expression whose operand is a variable of the corresponding tuple
type. No other leaf AST node can evaluate to a pack value.

A tuple cannot be indexed unless all of its components are scalar. There is no
syntax for indexing into a pack.

### Pattern semantics

Pack expansions can also appear in patterns. The semantics are chosen to follow
the general principle that pattern matching is the inverse of expression

> evaluation, so for example if the pattern `(..., [:]x: auto)` matches some

Reviewer: I thought we couldn't use `auto` with packs?

Author: The rule is that `auto` never deduces as a pack type, but we're not violating that here; `x` has to be a tuple or an array, or else we couldn't apply `[:]` to it.

value there is a corresponding pack value, and the other way around, but the two
values are not identical, because they support different operations. In
particular, there is no way to access a specific element of a pack, whereas
that's almost the only operation that tuples support.

We typically think of tuples as fixed-size sequences of elements, but in order
to support variadics, we need to be able to reason symbolically about packs and
tuples that contain subsequences of an indeterminate number of elements that
share a common structure. For example, in a function with deduced parameters
`[N:! SizeType, M:! SizeType, Ts:! [type; N], Us:! [type; M]]`, the type
`(..., [:]Ts, i32, ..., Optional([:]Us))` is a tuple whose elements are an
indeterminate number of types, followed by `i32`, followed by a different
indeterminate number of types that all have the form `Optional(U)` for some type
`U`. Consequently, we need to represent tuples and packs in a more general way
during symbolic evaluation, by saying that a pack or tuple consists of a
sequence of _pack components_.

A pack component consists of a value called the _representative_, and an arity,

> which may both be symbolic. As their name suggests, pack components cannot occur

Reviewer: ```suggestion
both of which may be symbolic. As their name suggests, pack components cannot occur
```

on their own, but only as part of an enclosing pack (or tuple). Pack components
have no literal syntax in Carbon, but for purposes of illustration we will use
the notation `<V, N>` to represent a pack component with representative V and
arity N. There is a special symbolic variable called the _pack index_, which can
only be used in the representative of a pack component, and only as the
subscript of an indexing expression on a tuple or statically sized array. The
pack index is used solely as a placeholder, and never has a value. We will use
`$I` to represent it for purposes of illustration. _Expanding_ a component C
with arity N (where N is a known constant) means replacing it with N components
that have arity 1, where the representative of the k'th such component is a copy
of the representative of C, where every usage of the pack index is replaced with
k-1.

For example, continuing the earlier example, the expression
`(..., [:]Ts, i32, ..., Optional([:]Us))` symbolically evaluates to

> `(<Ts[$I], N>, <i32, 1>, <Us[$I], M>)` (the language rules leading to this

Reviewer: Both pack components use `$I` instead of using different names, like `$I` and `$J`?

Author: Yep. There's only ever one pack index variable, so giving it different names in different contexts would be confusing. I know it seems weird for this to be effectively a global variable, but there are definitely cases where it's important that two separate occurrences of the pack index are the same when they are brought together, and I can't come up with any cases where treating them as the same causes a problem.

A pack component is _variadic_ if its arity is unknown, and _scalar_ if its
arity is known to be 1 and does not refer to the pack index. In contexts where
all pack components are known to be scalar, we will sometimes refer to them as
"elements". Pack values are always assumed to be normalized, meaning that every
component is either scalar or variadic. This is always possible because if the
component's arity is known but not 1, we can expand it, which produces a
sequence of scalar components by construction. The _shape_ of a pack is another
pack, found by replacing all representatives with `()` and leaving the arities
unchanged. A _pack type_ is the type of a pack, and can be represented as a pack
value whose representatives are types. The type of a pack is found by replacing
each of its representatives with the representative's type.

Even during symbolic evaluation, we need to maintain the structural equivalence
between packs and tuples, so in this more generalized model, a tuple value
consists of a sequence of pack components. Name bindings can have pack types,
but only if they are local to a single function. Note that by construction,

> bindings with pack types always have an explicit `[:]` operator in the type

Reviewer: ```suggestion
bindings with pack types always have an explicit `[:]` operator in their type
```

```cpp
template <ConvertibleToString... Ts>
std::string StrCat(const Ts&... params) {
  if constexpr (sizeof...(params) == 0) {
    return "";
  } else {
    std::string result;
    result.reserve((params.Length() + ... + 0));
    StrCatImpl(&result, params...);
    return result;
  }
}

template <ConvertibleToString T, ConvertibleToString... Ts>
void StrCatImpl(std::string* out, const T& first, const Ts&... rest) {
  out->append(first.ToString());
  if constexpr (sizeof...(rest) > 0) {
    StrCatImpl(out, rest...);
  }
}

> ```

Reviewer 4: @josh11b asked me whether I'd write / prefer the form with `if constexpr` (with two termination conditions) or the form with the extra function overload. I said I'd actually write this:
````suggestion
```cpp
template <ConvertibleToString... Ts>
std::string StrCat(const Ts&... params) {
  std::string result;
  result.reserve((params.Length() + ... + 0));
  (out->append(params.ToString()), ...);
  return result;
}
```
````

Author: Me too, but I don't think that's representative of what a typical C++ programmer would do in this situation, so this gets back to the question of whether we want the C++ examples to reflect typical usage or expert usage (see also [this comment thread](https://github.com/carbon-language/carbon-lang/pull/2240#discussion_r1715815035)).

very hard to see how we could define them as library types without circularity.
Packs have much the same problem, since a type representing a pack would itself
have to be a variadic type. The need for both types to deeply support symbolic
computation adds some major challenges here as well.

However, if tuples could ever be library types, they would certainly need to be
defined variadically, so in that sense this proposal may actually support that
principle. In addition, the fact that pack types are built-in is arguably

> mitigated by the fact that pack types cannot be explicitly named in Carbon code.

Reviewer: May also be worth saying that some important standard interfaces like `Call` have variadic parameters.

Author: I think the rationale is supposed to tie back to the project goals and design principles, not to other parts of the design.

Reviewer: This is part of https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/library_apis_only.md

Author: OK, I've made changes to explicitly address this. In the course of doing that, I've convinced myself that this proposal doesn't actually deviate from the library API principle, so I've revised this passage to reflect that.

Extending the type system to support deduction that splits into multiple cases
would add a fearsome amount of complexity to the type system.

> ##### Identifying potential matchings

Reviewer: Might be good to use some different formatting for the variables N, M, i, etc., like _italics_ or `code`.

Author: I don't really like either of those options, but how about this?

Our solution will rely on being able to identify which pattern pack elements can
potentially match which scrutinee pack elements. We can do so as follows:

We will refer to the elements of the pack pattern as "parameters", and the
elements of the scrutinee expression as "symbolic arguments". We will use the
term "actual arguments" to refer to the elements of the scrutinee after
monomorphization, so a single symbolic argument may correspond to any number of
actual arguments, including zero (but only if the expression is variadic). Note
that throughout this discussion, "parameters" and "arguments" are always types;
we will sometimes say "parameter type" or "argument type" when there is a risk
of confusion (FIXME there might still be too much risk of confusion, especially
because we shift to more conventional meanings for "argument" and "parameter" in
the next section). We will refer to arguments and parameters as "scalar" if they
have size 1, and "variadic" if they have indeterminate size (these are the only
possibilities, because packs are normalized).

> A pack pattern consists of N leading scalar parameters, optionally followed by a

Reviewer: I don't think the below addresses the case where there is no variadic parameter.

Author: I think it does, but I've added a note to the next paragraph to make it clearer.

a variadic type. Variadic types are themselves patterns, so we will break this
down into cases based on the possible forms of a pattern.

##### Tuple type

FIXME too much nesting

At type checking time, we don't necessarily know which parameters each argument
will match with, or the other way around. For example, consider the following
code:

```carbon

> fun F(a: i32, ..., [:]b: [i32;], c: i32);

Reviewer: ```suggestion
fn F(a: i32, ..., [:]b: [i32;], c: i32);
```

> fun G(..., [:]x: [i32;]) {

Reviewer: ```suggestion
fn G(..., [:]x: [i32;]) {
```

  F(1, 2, ..., [:]x);
}
```

If `x` is empty, the `2` will match with `c`, and otherwise the `2` will match
with an element of `b`. Similarly, if `x` is not empty, its last element will
match `c`, and the remaining elements (if any) will match elements of `b`.

> However, at type-checking time we don't know the size of `x` yes, so we don't

Reviewer: ```suggestion
However, at type-checking time we don't know the size of `x` yet, so we don't
```

know which will occur. On the other hand, the `1` will always match `a`.

> In general, We want type checking to fail if any possible monomorphization of

Reviewer: ```suggestion
In general, we want type checking to fail if any possible monomorphization of
```

the generic code would fail to typecheck. In this case that means we want type
checking to fail if any of the potential argument-parameter mappings could fail
to typecheck after monomorphization. Furthermore, for reasons of readability as
well as efficiency, we want type checking to fail if any two potential mappings
would deduce inconsistent values for any deduced parameter. However, in general
this is intractable, because the number of distinct ways to map argument

> expressions to parameters is exponential in the number of variadic arguments.

Reviewer: "exponential" here could use some justification (and I'm not sure it is precisely accurate).

Author: How's this?

I don't want to get too into the weeds here, because it feels pretty intuitive to me that anything that starts with "the number of distinct ways" probably isn't going to have a happy ending

Reviewer: Maybe it would be better to avoid "exponential" as too specific, and use the more generic https://en.wikipedia.org/wiki/Combinatorial_explosion term instead?

Author: I'm not sure how to use "combinatorial explosion" as an adjective, especially if I need to provide some justification.

> expressions to parameters is exponential in the number of variadic arguments.

Reviewer: ```suggestion
expressions to parameters is exponential in the number of variadic arguments in the worst case.
```

Introducing type deduction further complicates the situation. For example:

```carbon

> fun H[Ts:! [type;]](a: i32, ..., [:]b: Ts, c: String) -> (..., [:]Ts);

Reviewer: ```suggestion
fn H[Ts:! [type;]](a: i32, ..., [:]b: Ts, c: String) -> (..., [:]Ts);
```

external impl P as ImplicitAs(i32);
external impl Q as ImplicitAs(String);

> fun I(x: [i32;], y: [f32;], z: [String;]) {

Reviewer: ```suggestion
fn I(x: [i32;], y: [f32;], z: [String;]) {
```

>   result: auto = H(..., [:]x, {} as P, ..., [:]y, {} as Q, ..., [:]z);

Reviewer: ```suggestion
  var result: auto = H(..., [:]x, {} as P, ..., [:]y, {} as Q, ..., [:]z);
```

"before" and "after" reversed. And as noted earlier, central argument
expressions can only match the variadic parameter. In summary, we can identify
the possible matches for an argument expression E as follows:

-   If E is leading, let i be one more than the number of earlier non-variadic
    argument expressions:
    -   If E is non-variadic, and there are no earlier variadic argument
        expressions, then E can only match the i'th leading parameter.
    -   Otherwise, E can match the i'th leading parameter, any later leading
        parameters, and the variadic parameter.
-   If E is trailing, let i be one more than the number of later non-variadic
    argument expressions:
    -   If E is non-variadic, and there are no later variadic argument
        expressions, then E can only match the i'th trailing parameter from the
        end.
    -   Otherwise, E can match the i'th trailing parameter from the end, any
        earlier trailing parameters, and the variadic parameter.
-   Otherwise, E can only match the variadic parameter.

###### The type-checking algorithm

In order to avoid type deduction that splits into multiple cases, we require

> that if the variadic parameter has a deduced type that is used in more than one

Reviewer: Isn't this only when it a type pack? Or is the type only deduced when it is a pack?

Author: Yeah, I don't think there's any way to deduce a non-pack type for the variadic parameter. In principle I think it could have a type like `[T;]` where `T` is deduced, but I wouldn't call a type with a deduced component a "deduced type". I'm open to suggestions for how to make that clearer, though. In any event, for now we shouldn't allow that, since we don't allow its non-variadic counterpart (I've added a note about that up above).

Reviewer: I think it would be clearer if the text said the word "pack" to clarify that is what is being discussed. I don't think `T` can be deduced from a `[T;]` type expression, since it could have length 0. In the `Min` case, the `first` parameter has type `T` and can be used to ensure that `T` can be deduced from the arguments. It is a requirement that every deduced parameter is guaranteed to be deducible from the arguments (and not the return type). I think that means a parameter with a `[T;]` type must be paired with some other parameter that can be used to deduce `T`.

Author: I've tweaked this rule to be a bit more general, which I think makes this moot.

place (as `Ts` is in the earlier example of this problem), all variadic
arguments must have a known size (which means, as noted earlier, that we will
expand them into a sequence of non-variadic arguments for type-checking
purposes). This ensures that we can deduce the value of the array
element-by-element in the cases where we will need to use that deduced value.

> **Open question:** Is that restriction too strict? If so, it may be possible
> to forbid only situations that would actually cause type deduction to split
> into multiple cases. As well as being much less restrictive, that would avoid
> the need to give special treatment to deduced arrays that are used only once.
> It would still disallow cases like the call to `H` above, but that call seems
> unnatural for reasons that seem closely related to the fact that its type
> splits into cases.

To avoid the potential for exponential blow-up, we will use a much more

> tractable conservative approximation of the ideal algorithm. We type-check each

Reviewer: ```suggestion
tractable conservative approximation of the precise algorithm. We type-check each
```
